{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/work/clusterEvo/new_tests/eggNOG\n"
     ]
    }
   ],
   "source": [
    "from matplotlib import pyplot as plt\n",
    "import seaborn as sns\n",
    "import pandas  as pd\n",
    "import numpy   as np\n",
    "import os\n",
    "import re\n",
    "import itertools\n",
    "from scipy.spatial.distance import squareform\n",
    "from collections import Counter\n",
    "import ete3\n",
    "\n",
    "%run /work/jupyter_notebooks/gene\\ family\\ distances/correlate_evolution.ipynb\n",
    "%cd  /work/clusterEvo/new_tests/eggNOG/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly\n",
    "import chart_studio.plotly as ptl\n",
    "import plotly.graph_objects as go\n",
    "\n",
    "ptl.sign_in('lthiberiol', 'm15ikp59lt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "eggNOG_groups = pd.read_csv('2157_members.tsv',\n",
    "                            sep='\\t',\n",
    "                            header=None,\n",
    "                            usecols=[1,2,3,4],\n",
    "                            names=['group_id', 'num_proteins', 'num_taxa', 'members'])\n",
    "\n",
    "tmp           = eggNOG_groups.members.map(lambda cell: [int(taxid) \n",
    "                                                        for taxid in re.findall('(\\d+)\\.(?:[^,]+)', cell)])\n",
    "tmp.name      = 'taxa'\n",
    "eggNOG_groups = eggNOG_groups.join(tmp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(31, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>group_id</th>\n",
       "      <th>num_proteins</th>\n",
       "      <th>num_taxa</th>\n",
       "      <th>members</th>\n",
       "      <th>taxa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1392</th>\n",
       "      <td>arCOG00207</td>\n",
       "      <td>11</td>\n",
       "      <td>10</td>\n",
       "      <td>1104324.P186_0784,1104324.P186_1686,1198449.AC...</td>\n",
       "      <td>[1104324, 1104324, 1198449, 178306, 272557, 41...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1724</th>\n",
       "      <td>arCOG00599</td>\n",
       "      <td>11</td>\n",
       "      <td>10</td>\n",
       "      <td>178306.PAE0153,178306.PAE1775,368408.Tpen_1003...</td>\n",
       "      <td>[178306, 178306, 368408, 384616, 397948, 41035...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2260</th>\n",
       "      <td>arCOG01211</td>\n",
       "      <td>11</td>\n",
       "      <td>10</td>\n",
       "      <td>1104324.P186_0847,1365176.N186_06385,178306.PA...</td>\n",
       "      <td>[1104324, 1365176, 178306, 368408, 368408, 391...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3009</th>\n",
       "      <td>arCOG02069</td>\n",
       "      <td>11</td>\n",
       "      <td>10</td>\n",
       "      <td>1202768.JROF01000004_gene3179,1220534.B655_117...</td>\n",
       "      <td>[1202768, 1220534, 1227453, 192952, 269797, 41...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3449</th>\n",
       "      <td>arCOG02623</td>\n",
       "      <td>11</td>\n",
       "      <td>10</td>\n",
       "      <td>1006006.Mcup_0813,273057.SSO1595,273063.STK_20...</td>\n",
       "      <td>[1006006, 273057, 273063, 330779, 399549, 4535...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        group_id  num_proteins  num_taxa  \\\n",
       "1392  arCOG00207            11        10   \n",
       "1724  arCOG00599            11        10   \n",
       "2260  arCOG01211            11        10   \n",
       "3009  arCOG02069            11        10   \n",
       "3449  arCOG02623            11        10   \n",
       "\n",
       "                                                members  \\\n",
       "1392  1104324.P186_0784,1104324.P186_1686,1198449.AC...   \n",
       "1724  178306.PAE0153,178306.PAE1775,368408.Tpen_1003...   \n",
       "2260  1104324.P186_0847,1365176.N186_06385,178306.PA...   \n",
       "3009  1202768.JROF01000004_gene3179,1220534.B655_117...   \n",
       "3449  1006006.Mcup_0813,273057.SSO1595,273063.STK_20...   \n",
       "\n",
       "                                                   taxa  \n",
       "1392  [1104324, 1104324, 1198449, 178306, 272557, 41...  \n",
       "1724  [178306, 178306, 368408, 384616, 397948, 41035...  \n",
       "2260  [1104324, 1365176, 178306, 368408, 368408, 391...  \n",
       "3009  [1202768, 1220534, 1227453, 192952, 269797, 41...  \n",
       "3449  [1006006, 273057, 273063, 330779, 399549, 4535...  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "one_duplication = eggNOG_groups.query('(num_proteins == num_taxa+1) &'\n",
    "                                      '(num_taxa >= 10) &'\n",
    "                                      '(num_taxa <= 10)')\n",
    "# single_copy.to_csv('single_copy_genes.tsv', sep='\\t')\n",
    "print(one_duplication.shape)\n",
    "one_duplication.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def taxa_jaccard(ids):\n",
    "    taxon_set1 = set(one_duplication.loc[one_duplication.group_id==ids[0],\n",
    "                                         'taxa'].squeeze())\n",
    "    taxon_set2 = set(one_duplication.loc[one_duplication.group_id==ids[1],\n",
    "                                         'taxa'].squeeze())\n",
    "    \n",
    "    j_index = len(\n",
    "        taxon_set1.intersection(taxon_set2)\n",
    "    ) / len(\n",
    "        taxon_set1.union(taxon_set2)\n",
    "    )\n",
    "    \n",
    "    return(j_index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "group1, group2, j_indices = [], [], []\n",
    "for ids in itertools.combinations(one_duplication.group_id.values, 2):\n",
    "    j_indices.append(taxa_jaccard(ids))\n",
    "    group1.append(ids[0])\n",
    "    group2.append(ids[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "jaccard_df = pd.DataFrame(columns=[  'group1', 'group2', 'jaccard'],\n",
    "                          data   =zip(group1,   group2,   j_indices))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>group1</th>\n",
       "      <th>group2</th>\n",
       "      <th>jaccard</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>arCOG00599</td>\n",
       "      <td>arCOG05535</td>\n",
       "      <td>0.818182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>294</th>\n",
       "      <td>arCOG05433</td>\n",
       "      <td>arCOG05535</td>\n",
       "      <td>0.818182</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         group1      group2   jaccard\n",
       "41   arCOG00599  arCOG05535  0.818182\n",
       "294  arCOG05433  arCOG05535  0.818182"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "jaccard_df.query('jaccard >= 0.7')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_duplicated_taxa(group_id):\n",
    "    taxa = re.findall('(?:^|,)(\\d+?)\\.', \n",
    "                      one_duplication.loc[one_duplication.group_id==group_id,\n",
    "                                         'members'].squeeze(),\n",
    "                      re.M)\n",
    "    \n",
    "    for taxon, count in Counter(taxa).items():\n",
    "        if count > 1:\n",
    "            yield(taxon)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'178306'"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "next(get_duplicated_taxa('arCOG05535'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'178306'"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "next(get_duplicated_taxa('arCOG00599'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'178306'"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "next(get_duplicated_taxa('arCOG05433'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "eggNOG_trees = pd.read_csv('2157_trees.tsv',\n",
    "                           sep      ='\\t',\n",
    "                           header   =None,\n",
    "                           usecols  =[1,2,3],\n",
    "                           index_col=0,\n",
    "                           names    =['group_id', 'fast', 'tree'])\n",
    "eggNOG_trees = eggNOG_trees.reindex(index=one_duplication.group_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {},
   "outputs": [],
   "source": [
    "m1 = get_matrix_from_tree(\n",
    "    ete3.Tree(eggNOG_trees.loc['arCOG05535',\n",
    "                               'tree'])\n",
    ")\n",
    "\n",
    "m2 = get_matrix_from_tree(\n",
    "    ete3.Tree(eggNOG_trees.loc['arCOG00599',\n",
    "                               'tree'])\n",
    ")\n",
    "\n",
    "m3 = get_matrix_from_tree(\n",
    "    ete3.Tree(eggNOG_trees.loc['arCOG05433',\n",
    "                               'tree'])\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 446,
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix1 = m1.copy()\n",
    "matrix2 = m2.copy()\n",
    "\n",
    "tmp_taxa = []\n",
    "for index in matrix1.index:\n",
    "    genome, gene = re.search('^(\\d+?)\\.(.*)$', index).groups()\n",
    "    tmp_taxa.append([index, genome, gene])\n",
    "\n",
    "taxa1 = pd.DataFrame(columns=['taxon', 'genome', 'gene'],\n",
    "                     data   =tmp_taxa)\n",
    "\n",
    "tmp_taxa = []\n",
    "for index in matrix2.index:\n",
    "    genome, gene = re.search('^(\\d+?)\\.(.*)$', index).groups()\n",
    "    tmp_taxa.append([index, genome, gene])\n",
    "\n",
    "taxa2 = pd.DataFrame(columns=['taxon', 'genome', 'gene'],\n",
    "                     data   =tmp_taxa)\n",
    "\n",
    "shared_genomes = np.intersect1d(taxa1.genome.unique(), \n",
    "                                taxa2.genome.unique())\n",
    "\n",
    "taxa1 = taxa1[taxa1.genome.isin(shared_genomes)]\n",
    "taxa2 = taxa2[taxa2.genome.isin(shared_genomes)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 447,
   "metadata": {},
   "outputs": [],
   "source": [
    "taxa1_frequency = Counter(taxa1.genome) \n",
    "taxa2_frequency = Counter(taxa2.genome)\n",
    "\n",
    "for genome in shared_genomes:\n",
    "    genome1_count = taxa1_frequency[genome]\n",
    "    genome2_count = taxa2_frequency[genome]\n",
    "\n",
    "    if genome1_count > 1:\n",
    "        for _ in range(genome1_count - 1):\n",
    "            for index, row in taxa2.iloc[::-1].query('genome == @genome').iterrows():\n",
    "                row.taxon += f'.{_}'\n",
    "                taxa2 = taxa2.append(row, ignore_index=True)\n",
    "                \n",
    "                reference_name = re.sub('\\.\\d+$', '', row.taxon, flags=re.M)\n",
    "                matrix2[    row.taxon] = matrix2[    reference_name]\n",
    "                matrix2.loc[row.taxon] = matrix2.loc[reference_name]\n",
    "\n",
    "\n",
    "    if genome2_count > 1:\n",
    "        for _ in range(genome2_count - 1):\n",
    "            for index, row in taxa1.query('genome == @genome').iterrows():\n",
    "                row.taxon += f'.{_}'\n",
    "                taxa1 = taxa1.append(row, ignore_index=True)\n",
    "                \n",
    "                reference_name = re.sub('\\.\\d+$', '', row.taxon, flags=re.M)\n",
    "                matrix1[    row.taxon] = matrix1[    reference_name]\n",
    "                matrix1.loc[row.taxon] = matrix1.loc[reference_name]\n",
    "    \n",
    "taxa1.sort_values('genome', inplace=True)\n",
    "taxa2.sort_values('genome', inplace=True)\n",
    "\n",
    "matrix1 = matrix1.reindex(index  =taxa1.taxon, \n",
    "                          columns=taxa1.taxon, \n",
    "                          copy   =True)\n",
    "matrix2 = matrix2.reindex(index  =taxa2.taxon, \n",
    "                          columns=taxa2.taxon, \n",
    "                          copy   =True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 603,
   "metadata": {},
   "outputs": [],
   "source": [
    "triu_indices = np.triu_indices_from(matrix1, k=1)\n",
    "condensed1   = matrix1.values[triu_indices]\n",
    "condensed2   = matrix2.values[triu_indices]\n",
    "\n",
    "#\n",
    "# ignore comparisons within the same species\n",
    "matrix1_distinct_genomes = (taxa1.iloc[triu_indices[0], 1].values != \\\n",
    "                            taxa1.iloc[triu_indices[1], 1].values\n",
    "                           ).astype(int)\n",
    "matrix2_distinct_genomes = (taxa2.iloc[triu_indices[0], 1].values !=\\\n",
    "                            taxa2.iloc[triu_indices[1], 1].values\n",
    "                           ).astype(int)\n",
    "\n",
    "model = Model(line)\n",
    "data  = Data(condensed1, \n",
    "             condensed2, \n",
    "             wd=matrix1_distinct_genomes, \n",
    "             we=matrix2_distinct_genomes\n",
    ")\n",
    "odr = ODR(data, \n",
    "          model,\n",
    "          beta0=[np.std(condensed2) /\\\n",
    "                 np.std(condensed1)]\n",
    "         )\n",
    "\n",
    "regression = odr.run()\n",
    "\n",
    "residual_df = pd.DataFrame(columns=['x_taxon1',   'x_genome1', \n",
    "                                    'x_taxon2',   'x_genome2', \n",
    "                                    \n",
    "                                    'y_taxon1',   'y_genome1', \n",
    "                                    'y_taxon2',   'y_genome2', \n",
    "                                    \n",
    "                                    'x_residual', 'y_residual'],\n",
    "                           data   =zip(matrix1.index[triu_indices[0]],\n",
    "                                       taxa1.iloc[triu_indices[0], 1].values,\n",
    "                                       matrix1.index[triu_indices[1]],\n",
    "                                       taxa1.iloc[triu_indices[1], 1].values,\n",
    "                                       \n",
    "                                       matrix2.index[triu_indices[0]],\n",
    "                                       taxa2.iloc[triu_indices[0], 1].values,\n",
    "                                       matrix2.index[triu_indices[1]],\n",
    "                                       taxa2.iloc[triu_indices[1], 1].values,\n",
    "                                       \n",
    "                                       abs(regression.delta),\n",
    "                                       abs(regression.eps))\n",
    "                          )\n",
    "residual_df['residual_total'] = residual_df.x_residual + residual_df.y_residual\n",
    "\n",
    "within_genomes = ((residual_df.x_genome1 == residual_df.x_genome2) | \n",
    "                  (residual_df.y_genome1 == residual_df.y_genome2))\n",
    "\n",
    "residual_df.drop(index=residual_df.index[within_genomes], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 613,
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix1_homologs = taxa1.loc[taxa1.genome==\"178306\", \n",
    "                           'taxon'].values\n",
    "matrix2_homologs = taxa2.loc[taxa2.genome==\"178306\", \n",
    "                           'taxon'].values\n",
    "\n",
    "homolog_combinations = pd.DataFrame(columns=['homolog1', \n",
    "                                             'homolog2', \n",
    "                                             'residual_sum'])\n",
    "for homolog1, homolog2 in itertools.product(matrix1_homologs,\n",
    "                                            matrix2_homologs):\n",
    "    tmp_df = residual_df.query('(x_taxon1 == @homolog1 | x_taxon2 == @homolog1) &'\n",
    "                               '(y_taxon1 == @homolog2 | y_taxon2 == @homolog2)')\n",
    "    \n",
    "    if not tmp_df.shape[0]:\n",
    "        continue\n",
    "    \n",
    "    homolog1 = re.sub('\\.\\d$', \n",
    "                      '',\n",
    "                      homolog1, \n",
    "                      flags=re.M)\n",
    "    homolog2 = re.sub('\\.\\d$',\n",
    "                      '', \n",
    "                      homolog2, \n",
    "                      flags=re.M)\n",
    "    \n",
    "    homolog_combinations = homolog_combinations.append(\n",
    "        pd.Series(data=[homolog1, \n",
    "                        homolog2, \n",
    "                        tmp_df.residual_total.sum()],\n",
    "                  index=['homolog1', \n",
    "                         'homolog2', \n",
    "                         'residual_sum']), \n",
    "        ignore_index=True\n",
    "    )\n",
    "    \n",
    "homolog_combinations.sort_values('residual_sum', inplace=True)\n",
    "best_pairs = set()\n",
    "while homolog_combinations.shape[0]:\n",
    "    first_row = homolog_combinations.iloc[0]\n",
    "    best_pairs.add((first_row.homolog1, first_row.homolog2))\n",
    "    homolog_combinations = homolog_combinations.query(f'(homolog1 != \"{first_row.homolog1}\") & '\n",
    "                                                      f'(homolog2 != \"{first_row.homolog2}\")').copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def match_copies(matrix1, matrix2, taxa1, taxa2):\n",
    "\n",
    "    triu_indices = np.triu_indices_from(matrix1, k=1)\n",
    "    condensed1   = matrix1.values[triu_indices]\n",
    "    condensed2   = matrix2.values[triu_indices]\n",
    "\n",
    "    #\n",
    "    # ignore comparisons within the same species\n",
    "    matrix1_distinct_genomes = (taxa1.iloc[triu_indices[0], 1].values != \\\n",
    "                                taxa1.iloc[triu_indices[1], 1].values\n",
    "                               ).astype(int)\n",
    "    matrix2_distinct_genomes = (taxa2.iloc[triu_indices[0], 1].values !=\\\n",
    "                                taxa2.iloc[triu_indices[1], 1].values\n",
    "                               ).astype(int)\n",
    "\n",
    "    model = Model(line)\n",
    "    data  = Data(condensed1, \n",
    "                 condensed2, \n",
    "                 wd=matrix1_distinct_genomes, \n",
    "                 we=matrix2_distinct_genomes\n",
    "    )\n",
    "    odr = ODR(data, \n",
    "              model,\n",
    "              beta0=[np.std(condensed2) /\\\n",
    "                     np.std(condensed1)]\n",
    "             )\n",
    "\n",
    "    regression = odr.run()\n",
    "\n",
    "    residual_df = pd.DataFrame(columns=['x_taxon1',   'x_genome1', \n",
    "                                        'x_taxon2',   'x_genome2', \n",
    "\n",
    "                                        'y_taxon1',   'y_genome1', \n",
    "                                        'y_taxon2',   'y_genome2', \n",
    "\n",
    "                                        'x_residual', 'y_residual'],\n",
    "                               data   =zip(matrix1.index[triu_indices[0]],\n",
    "                                           taxa1.iloc[triu_indices[0], 1].values,\n",
    "                                           matrix1.index[triu_indices[1]],\n",
    "                                           taxa1.iloc[triu_indices[1], 1].values,\n",
    "\n",
    "                                           matrix2.index[triu_indices[0]],\n",
    "                                           taxa2.iloc[triu_indices[0], 1].values,\n",
    "                                           matrix2.index[triu_indices[1]],\n",
    "                                           taxa2.iloc[triu_indices[1], 1].values,\n",
    "\n",
    "                                           abs(regression.delta),\n",
    "                                           abs(regression.eps))\n",
    "                              )\n",
    "    residual_df['residual_total'] = residual_df.x_residual + residual_df.y_residual\n",
    "\n",
    "    within_genomes = ((residual_df.x_genome1 == residual_df.x_genome2) | \n",
    "                      (residual_df.y_genome1 == residual_df.y_genome2))\n",
    "\n",
    "    residual_df.drop(index=residual_df.index[within_genomes], inplace=True)\n",
    "    \n",
    "    matrix1_homologs = taxa1.loc[taxa1.genome==\"178306\", \n",
    "                               'taxon'].values\n",
    "    matrix2_homologs = taxa2.loc[taxa2.genome==\"178306\", \n",
    "                               'taxon'].values\n",
    "\n",
    "    homolog_combinations = pd.DataFrame(columns=['homolog1', \n",
    "                                                 'homolog2', \n",
    "                                                 'residual_sum'])\n",
    "    for homolog1, homolog2 in itertools.product(matrix1_homologs,\n",
    "                                                matrix2_homologs):\n",
    "        tmp_df = residual_df.query('(x_taxon1 == @homolog1 | x_taxon2 == @homolog1) &'\n",
    "                                   '(y_taxon1 == @homolog2 | y_taxon2 == @homolog2)')\n",
    "\n",
    "        if not tmp_df.shape[0]:\n",
    "            continue\n",
    "\n",
    "        homolog1 = re.sub('\\.\\d$', \n",
    "                          '',\n",
    "                          homolog1, \n",
    "                          flags=re.M)\n",
    "        homolog2 = re.sub('\\.\\d$',\n",
    "                          '', \n",
    "                          homolog2, \n",
    "                          flags=re.M)\n",
    "\n",
    "        homolog_combinations = homolog_combinations.append(\n",
    "            pd.Series(data=[homolog1, \n",
    "                            homolog2, \n",
    "                            tmp_df.residual_total.sum()],\n",
    "                      index=['homolog1', \n",
    "                             'homolog2', \n",
    "                             'residual_sum']), \n",
    "            ignore_index=True\n",
    "        )\n",
    "\n",
    "    homolog_combinations.sort_values('residual_sum', inplace=True)\n",
    "    best_pairs = set()\n",
    "    while homolog_combinations.shape[0]:\n",
    "        first_row = homolog_combinations.iloc[0]\n",
    "        best_pairs.add((first_row.homolog1, first_row.homolog2))\n",
    "        homolog_combinations = homolog_combinations.query(f'(homolog1 != \"{first_row.homolog1}\") & '\n",
    "                                                          f'(homolog2 != \"{first_row.homolog2}\")').copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 455,
   "metadata": {},
   "outputs": [],
   "source": [
    "triu_indices = np.triu_indices_from(matrix1, k=1)\n",
    "condensed1   = matrix1.values[triu_indices]\n",
    "condensed2   = matrix2.values[triu_indices]\n",
    "\n",
    "matrix1_label = [f'{index1} VS {index2}' for index1, index2 in zip(matrix1.index[triu_indices[0]],\n",
    "                                                                   matrix1.index[triu_indices[1]])]\n",
    "matrix2_label = [f'{index1} VS {index2}' for index1, index2 in zip(matrix2.index[triu_indices[0]],\n",
    "                                                                   matrix2.index[triu_indices[1]])]\n",
    "\n",
    "combined_labels = [f'{label1}<br>{label2}' for label1, label2 in zip(matrix1_label,\n",
    "                                                                   matrix2_label)]\n",
    "\n",
    "node_color = ['red' if '178306' in label else 'black' for label in combined_labels]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 456,
   "metadata": {},
   "outputs": [],
   "source": [
    "node_color = []\n",
    "for label in matrix1_label:\n",
    "    \n",
    "    if label.count('178306') > 1:\n",
    "        node_color.append('blue')\n",
    "    elif '178306.PAE0154'  in label:\n",
    "        node_color.append('red')\n",
    "    elif '178306.PAE1775a' in label:\n",
    "        node_color.append('green')\n",
    "    else:\n",
    "        node_color.append('black')\n",
    "        \n",
    "node_shape = []\n",
    "for label in matrix2_label:\n",
    "    \n",
    "    if label.count('178306') > 1:\n",
    "        node_shape.append('diamond')\n",
    "    elif '178306.PAE0153'  in label:\n",
    "        node_shape.append('cross')\n",
    "    elif '178306.PAE1775' in label:\n",
    "        node_shape.append('triangle-up')\n",
    "    else:\n",
    "        node_shape.append('circle')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 457,
   "metadata": {},
   "outputs": [],
   "source": [
    "node_x      = condensed1.copy()\n",
    "node_y      = condensed2.copy()\n",
    "node_text   = combined_labels\n",
    "    \n",
    "node_trace = go.Scatter(\n",
    "    x        =node_x, \n",
    "    y        =node_y,\n",
    "    text     =node_text,\n",
    "    mode     ='markers',\n",
    "    name     ='Proteins',\n",
    "    hoverinfo='text',\n",
    "    opacity  =0.7,\n",
    "    marker   =dict(symbol      =node_shape,\n",
    "                   color       =node_color,\n",
    "                   size        =10,\n",
    "                   line_width  =1,\n",
    "                   line_color  ='white')\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 458,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/thiberio/test.html'"
      ]
     },
     "execution_count": 458,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fig = go.Figure(data=[node_trace], # here is important to add edges first so they don't cover nodes\n",
    "                layout=go.Layout(template          ='simple_white',\n",
    "                                 title             ='Test',\n",
    "                                 titlefont_size    =16,\n",
    "                                 showlegend        =True,\n",
    "                                 legend_orientation='h',\n",
    "                                 hovermode         ='closest',\n",
    "                                 margin            =dict(b=20,\n",
    "                                                         l=5,\n",
    "                                                         r=5,\n",
    "                                                         t=40),\n",
    "#                                  xaxis             =dict(showgrid     =False,\n",
    "#                                                          zeroline     =False,\n",
    "#                                                          showticklabels=False),\n",
    "#                                  yaxis             =dict(showgrid      =False, \n",
    "#                                                          zeroline      =False, \n",
    "#                                                          showticklabels=False)\n",
    "                                )\n",
    "                )\n",
    "\n",
    "plotly.offline.plot(fig, filename='/Users/thiberio/test.html', config={'scrollZoom': True}, auto_open=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 428,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>taxon</th>\n",
       "      <th>178306.PAE0154</th>\n",
       "      <th>178306.PAE1775a</th>\n",
       "      <th>384616.Pisl_1556</th>\n",
       "      <th>397948.Cmaq_0962</th>\n",
       "      <th>410359.Pcal_1246</th>\n",
       "      <th>444157.Tneu_0584</th>\n",
       "      <th>572478.Vdis_1352</th>\n",
       "      <th>698757.Pogu_1576</th>\n",
       "      <th>768679.TTX_1300</th>\n",
       "      <th>985053.VMUT_2172</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>taxon</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>178306.PAE0154</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.412437</td>\n",
       "      <td>2.558183</td>\n",
       "      <td>0.943222</td>\n",
       "      <td>2.288786</td>\n",
       "      <td>2.567279</td>\n",
       "      <td>2.187575</td>\n",
       "      <td>2.445332</td>\n",
       "      <td>2.526642</td>\n",
       "      <td>2.267118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178306.PAE1775a</th>\n",
       "      <td>2.412437</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.012768</td>\n",
       "      <td>3.022081</td>\n",
       "      <td>0.743371</td>\n",
       "      <td>1.021864</td>\n",
       "      <td>1.421092</td>\n",
       "      <td>0.265907</td>\n",
       "      <td>0.981227</td>\n",
       "      <td>1.500635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>384616.Pisl_1556</th>\n",
       "      <td>2.558183</td>\n",
       "      <td>1.012768</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.167827</td>\n",
       "      <td>0.848433</td>\n",
       "      <td>0.130635</td>\n",
       "      <td>1.566838</td>\n",
       "      <td>1.045663</td>\n",
       "      <td>0.395001</td>\n",
       "      <td>1.646381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>397948.Cmaq_0962</th>\n",
       "      <td>0.943222</td>\n",
       "      <td>3.022081</td>\n",
       "      <td>3.167827</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.898430</td>\n",
       "      <td>3.176923</td>\n",
       "      <td>2.797219</td>\n",
       "      <td>3.054976</td>\n",
       "      <td>3.136286</td>\n",
       "      <td>2.876762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>410359.Pcal_1246</th>\n",
       "      <td>2.288786</td>\n",
       "      <td>0.743371</td>\n",
       "      <td>0.848433</td>\n",
       "      <td>2.898430</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.857530</td>\n",
       "      <td>1.297441</td>\n",
       "      <td>0.776266</td>\n",
       "      <td>0.816892</td>\n",
       "      <td>1.376984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>444157.Tneu_0584</th>\n",
       "      <td>2.567279</td>\n",
       "      <td>1.021864</td>\n",
       "      <td>0.130635</td>\n",
       "      <td>3.176923</td>\n",
       "      <td>0.857530</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.575934</td>\n",
       "      <td>1.054759</td>\n",
       "      <td>0.404098</td>\n",
       "      <td>1.655477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>572478.Vdis_1352</th>\n",
       "      <td>2.187575</td>\n",
       "      <td>1.421092</td>\n",
       "      <td>1.566838</td>\n",
       "      <td>2.797219</td>\n",
       "      <td>1.297441</td>\n",
       "      <td>1.575935</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.453987</td>\n",
       "      <td>1.535297</td>\n",
       "      <td>0.379897</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>698757.Pogu_1576</th>\n",
       "      <td>2.445332</td>\n",
       "      <td>0.265907</td>\n",
       "      <td>1.045663</td>\n",
       "      <td>3.054976</td>\n",
       "      <td>0.776266</td>\n",
       "      <td>1.054759</td>\n",
       "      <td>1.453987</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.014122</td>\n",
       "      <td>1.533530</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>768679.TTX_1300</th>\n",
       "      <td>2.526642</td>\n",
       "      <td>0.981227</td>\n",
       "      <td>0.395001</td>\n",
       "      <td>3.136286</td>\n",
       "      <td>0.816892</td>\n",
       "      <td>0.404098</td>\n",
       "      <td>1.535297</td>\n",
       "      <td>1.014122</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.614840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>985053.VMUT_2172</th>\n",
       "      <td>2.267118</td>\n",
       "      <td>1.500635</td>\n",
       "      <td>1.646381</td>\n",
       "      <td>2.876762</td>\n",
       "      <td>1.376984</td>\n",
       "      <td>1.655477</td>\n",
       "      <td>0.379897</td>\n",
       "      <td>1.533530</td>\n",
       "      <td>1.614840</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "taxon             178306.PAE0154  178306.PAE1775a  384616.Pisl_1556  \\\n",
       "taxon                                                                 \n",
       "178306.PAE0154          0.000000         2.412437          2.558183   \n",
       "178306.PAE1775a         2.412437         0.000000          1.012768   \n",
       "384616.Pisl_1556        2.558183         1.012768          0.000000   \n",
       "397948.Cmaq_0962        0.943222         3.022081          3.167827   \n",
       "410359.Pcal_1246        2.288786         0.743371          0.848433   \n",
       "444157.Tneu_0584        2.567279         1.021864          0.130635   \n",
       "572478.Vdis_1352        2.187575         1.421092          1.566838   \n",
       "698757.Pogu_1576        2.445332         0.265907          1.045663   \n",
       "768679.TTX_1300         2.526642         0.981227          0.395001   \n",
       "985053.VMUT_2172        2.267118         1.500635          1.646381   \n",
       "\n",
       "taxon             397948.Cmaq_0962  410359.Pcal_1246  444157.Tneu_0584  \\\n",
       "taxon                                                                    \n",
       "178306.PAE0154            0.943222          2.288786          2.567279   \n",
       "178306.PAE1775a           3.022081          0.743371          1.021864   \n",
       "384616.Pisl_1556          3.167827          0.848433          0.130635   \n",
       "397948.Cmaq_0962          0.000000          2.898430          3.176923   \n",
       "410359.Pcal_1246          2.898430          0.000000          0.857530   \n",
       "444157.Tneu_0584          3.176923          0.857530          0.000000   \n",
       "572478.Vdis_1352          2.797219          1.297441          1.575935   \n",
       "698757.Pogu_1576          3.054976          0.776266          1.054759   \n",
       "768679.TTX_1300           3.136286          0.816892          0.404098   \n",
       "985053.VMUT_2172          2.876762          1.376984          1.655477   \n",
       "\n",
       "taxon             572478.Vdis_1352  698757.Pogu_1576  768679.TTX_1300  \\\n",
       "taxon                                                                   \n",
       "178306.PAE0154            2.187575          2.445332         2.526642   \n",
       "178306.PAE1775a           1.421092          0.265907         0.981227   \n",
       "384616.Pisl_1556          1.566838          1.045663         0.395001   \n",
       "397948.Cmaq_0962          2.797219          3.054976         3.136286   \n",
       "410359.Pcal_1246          1.297441          0.776266         0.816892   \n",
       "444157.Tneu_0584          1.575934          1.054759         0.404098   \n",
       "572478.Vdis_1352          0.000000          1.453987         1.535297   \n",
       "698757.Pogu_1576          1.453987          0.000000         1.014122   \n",
       "768679.TTX_1300           1.535297          1.014122         0.000000   \n",
       "985053.VMUT_2172          0.379897          1.533530         1.614840   \n",
       "\n",
       "taxon             985053.VMUT_2172  \n",
       "taxon                               \n",
       "178306.PAE0154            2.267118  \n",
       "178306.PAE1775a           1.500635  \n",
       "384616.Pisl_1556          1.646381  \n",
       "397948.Cmaq_0962          2.876762  \n",
       "410359.Pcal_1246          1.376984  \n",
       "444157.Tneu_0584          1.655477  \n",
       "572478.Vdis_1352          0.379897  \n",
       "698757.Pogu_1576          1.533530  \n",
       "768679.TTX_1300           1.614840  \n",
       "985053.VMUT_2172          0.000000  "
      ]
     },
     "execution_count": 428,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matrix1"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
