{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/nobackup1b/users/thiberio/eggNOG\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "import itertools\n",
    "import numpy as np\n",
    "from matplotlib import pyplot as plt\n",
    "import seaborn as sns\n",
    "import igraph as ig\n",
    "from IPython.display import HTML\n",
    "import multiprocessing\n",
    "import pickle as pkl\n",
    "import random\n",
    "from collections import Counter\n",
    "import os\n",
    "import subprocess\n",
    "from scipy.stats import mannwhitneyu\n",
    "import ete3\n",
    "from copy import deepcopy\n",
    "from math import ceil\n",
    "\n",
    "ncbi = ete3.NCBITaxa()\n",
    "%run assess_connections-functions.ipynb\n",
    "\n",
    "%cd ~/work/eggNOG/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "sampled_genomes = pd.read_csv('../kelsey/genomes.tab',\n",
    "                              sep='\\t',\n",
    "                              index_col=0)\n",
    "\n",
    "lineages = pd.DataFrame()\n",
    "for taxid in sampled_genomes.species_taxid.unique():\n",
    "    if pd.isna(taxid):\n",
    "        continue\n",
    "    lineages = lineages.append({tax_rank: tmp_taxid \n",
    "                                 for tmp_taxid, tax_rank in ncbi.get_rank(ncbi.get_lineage(taxid)).items()},\n",
    "                                ignore_index=True)\n",
    "lineages = lineages.reindex(columns=['class', 'family',  'genus', 'phylum',\n",
    "                                     'order', 'species', 'superkingdom']).copy()\n",
    "lineages = lineages.query('superkingdom == 2').copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "sampled_phyla = [phylum for phylum in lineages.phylum.unique().astype(int) if phylum > 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "working_groups  = pd.read_parquet('working_eggNOG_groups.parquet', engine='fastparquet')\n",
    "working_trees   = pd.read_parquet('working_eggNOG_trees.parquet' , engine='fastparquet')\n",
    "eggNOG_taxonomy = pd.read_parquet('eggNOG_taxonomy.parquet'      , engine='fastparquet')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('all_results.pkl', 'rb') as _:\n",
    "    phylum_connections = pkl.load(_)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "regular_connections     = []\n",
    "significant_connections = []\n",
    "\n",
    "for group_id, group_clusters in phylum_connections:\n",
    "    if not group_clusters:\n",
    "        continue\n",
    "    \n",
    "    for cluster_num, cluster_data in group_clusters.items():\n",
    "\n",
    "        for ref_phylum, phyla_dists in cluster_data.items():\n",
    "            \n",
    "            if not phyla_dists['df'].shape[0]:\n",
    "                continue\n",
    "\n",
    "            closest_phylum = phyla_dists['df'].iloc[0, 0]\n",
    "            if not closest_phylum in sampled_phyla:\n",
    "                continue\n",
    "\n",
    "            if phyla_dists['significant']:\n",
    "                significant_connections.append( (ref_phylum,\n",
    "                                                 phyla_dists['df'].iloc[0, 0],\n",
    "                                                 phyla_dists['df'].iloc[0, 1], \n",
    "                                                 '%s#%i' % (group_id, cluster_num)) )\n",
    "\n",
    "                for index, tmp_series in phyla_dists['df'].iloc[1:, :].iterrows():\n",
    "                    if tmp_series.phylum in sampled_phyla:\n",
    "                        regular_connections.append( (ref_phylum,\n",
    "                                                         tmp_series.phylum,\n",
    "                                                         tmp_series['median'], \n",
    "                                                         '%s#%i' % (group_id, cluster_num)) )\n",
    "\n",
    "            else:\n",
    "                for index, tmp_series in phyla_dists['df'].iterrows():\n",
    "                    if tmp_series.phylum in sampled_phyla:\n",
    "                        regular_connections.append( (ref_phylum,\n",
    "                                                         tmp_series.phylum,\n",
    "                                                         tmp_series['median'], \n",
    "                                                         '%s#%i' % (group_id, cluster_num)) )\n",
    "\n",
    "significant_df = pd.DataFrame(data=significant_connections, columns=['source', 'target', 'distance', 'group'])\n",
    "regular_df     = pd.DataFrame(data=regular_connections,     columns=['source', 'target', 'distance', 'group'])\n",
    "\n",
    "significant_df.dropna(how='any', inplace=True)\n",
    "regular_df.dropna(    how='any', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# fastTree\n",
    "#\n",
    "\n",
    "def assess_transfers_between_phyla(related_phyla):\n",
    "# related_phyla = (1090, 1117)\n",
    "\n",
    "    candidates    = set(significant_df.query('source==\"%i\" and target==\"%i\"' % related_phyla).group.values).union(\n",
    "                        significant_df.query('target==\"%i\" and source==\"%i\"' % related_phyla).group.values\n",
    "                    )\n",
    "    descriptions  = {f'from {related_phyla[0]} to {related_phyla[1]}': set(),\n",
    "                     f'from {related_phyla[1]} to {related_phyla[0]}': set(),\n",
    "                      'sisters':                                       set(),\n",
    "                      'no clear relationship between phyla':           set()}\n",
    "\n",
    "    for candidate in candidates:\n",
    "\n",
    "        group_id, cluster_num = candidate.split('#')\n",
    "\n",
    "        with cd('candidates/trees/'):\n",
    "            #\n",
    "            # test if there is rooted version of the tree (mad output) \n",
    "            if not os.path.isfile('%s-cluster%s.fastTree.rooted' % (group_id, cluster_num)):\n",
    "                continue\n",
    "\n",
    "            #\n",
    "            # we need the original tree cause mad removes support values, so we just transfer root positions\n",
    "            try:\n",
    "                tmp_tree   = match_rooting(ete3.Tree('%s-cluster%s.fastTree.rooted' % (group_id, cluster_num)),\n",
    "                                           ete3.Tree('%s-cluster%s.fastTree'        % (group_id, cluster_num)))\n",
    "            except ete3.parser.newick.NewickError:\n",
    "                continue\n",
    "\n",
    "        #\n",
    "        # extract taxIDs to subsample taxonomy tables\n",
    "        taxids = {}\n",
    "        for leaf in tmp_tree.get_leaf_names():\n",
    "            #\n",
    "            # leaf names are composed by <taxid>.<locus_tag>\n",
    "            #   ps: locus_tag may also have \".\" within it\n",
    "            tmp_taxid = int(leaf.split('.')[0])\n",
    "\n",
    "            if tmp_taxid not in taxids:\n",
    "                taxids[tmp_taxid] = []\n",
    "            taxids[tmp_taxid].append(leaf)\n",
    "\n",
    "        #\n",
    "        # select taxIDs from each assessed phylum...\n",
    "        phylum1_taxonomy = eggNOG_taxonomy.loc[taxids].query('phylum==@related_phyla[0]')\n",
    "        phylum2_taxonomy = eggNOG_taxonomy.loc[taxids].query('phylum==@related_phyla[1]')\n",
    "\n",
    "        #\n",
    "        # ... and their respect leaves\n",
    "        phylum1_leaves = set()\n",
    "        for taxid in phylum1_taxonomy.index:\n",
    "            phylum1_leaves.update(taxids[taxid])\n",
    "\n",
    "        phylum2_leaves = set()\n",
    "        for taxid in phylum2_taxonomy.index:\n",
    "            phylum2_leaves.update(taxids[taxid])\n",
    "        \n",
    "        all_leaves = phylum1_leaves.union(phylum2_leaves)\n",
    "\n",
    "        #\n",
    "        # as tree traversing through ete3 isn't very efficient, and doesn't scalate very well,\n",
    "        #   we create an iGraph dag for more effient traversing\n",
    "        dag  = tree_to_dag(tmp_tree)          # we need a directed version...\n",
    "        udag = dag.as_undirected(mode='each') # ... and an undirected one for different processes\n",
    "\n",
    "        #\n",
    "        # placeholders where we will add monophyletic clades for each phylum\n",
    "        phylum1_clades = set() \n",
    "        phylum2_clades = set()\n",
    "\n",
    "        ignored_nodes  = [] # descendants of monophyletic nodes should be ignored once their\n",
    "                            #   ancestors have been stored.\n",
    "\n",
    "        #\n",
    "        # traverse through internal nodes\n",
    "        for node in dag.vs:\n",
    "            if node.index in ignored_nodes:\n",
    "                continue\n",
    "\n",
    "            node_leaves = get_leaf_names(node)\n",
    "\n",
    "            if all_leaves.isdisjoint(node_leaves):\n",
    "                ignored_nodes.extend(get_descendant_indices(node, leaves=True))\n",
    "\n",
    "            #\n",
    "            # if there aren't leaves from other phyla within this node it is monophyletic\n",
    "            if phylum1_leaves.intersection(node_leaves) and phylum1_leaves.issuperset(node_leaves):\n",
    "                phylum1_clades.add(node.index)\n",
    "                ignored_nodes.extend(get_descendant_indices(node, leaves=True))\n",
    "\n",
    "            elif phylum2_leaves.intersection(node_leaves) and phylum2_leaves.issuperset(node_leaves):\n",
    "                phylum2_clades.add(node.index)\n",
    "                ignored_nodes.extend(get_descendant_indices(node, leaves=True))\n",
    "\n",
    "        #\n",
    "        # now we add some flexibility to the monophyly of nodes for three reasons:\n",
    "        #   1) donor nodes within gene trees will never be monophyletic since the recipient\n",
    "        #      must be nested within it.\n",
    "        #   2) if there are other transfers from the donor and/or recipient phyla to a 3rd one\n",
    "        #      we still want to capture it.\n",
    "        #   3) good ol' phylogenetic uncertainty\n",
    "        phylum1_clades = merge_polyphyletic_clades(phylum1_clades, udag)\n",
    "        phylum2_clades = merge_polyphyletic_clades(phylum2_clades, udag)\n",
    "\n",
    "        #\n",
    "        # flag if we can identify relations between phyla:\n",
    "        #   1 nested within 2\n",
    "        #   2 nested within 1\n",
    "        #   1 and 2 are sisters\n",
    "        #\n",
    "        # if no relation is identified, leave as false\n",
    "        phyla_relationship_flag = 0\n",
    "\n",
    "        for clade1, clade2 in itertools.product(phylum1_clades, phylum2_clades):\n",
    "\n",
    "            #\n",
    "            # capture ancestors of clade1 by querying nodes between itself and the root node\n",
    "            clade1_ancestors = udag.vs[clade1].get_shortest_paths(udag.vs[0])[0][1:]\n",
    "            #\n",
    "            # if clade2 within clade1 ancestors it means that clade1 is nested within clade2\n",
    "            #   evidence of transfer from clade2 -> clade1\n",
    "            if clade2 in clade1_ancestors:\n",
    "#                 support = dag.vs[clade2].in_edges()[0]['support']\n",
    "#                 print('%i is nested within %i             (%s)' % (related_phyla[0], related_phyla[1], candidate), support)\n",
    "                recipient_taxon         = fine_tune_recipient(dag.vs[clade1], \n",
    "                                                              related_phyla[0])\n",
    "                donor_taxon             = fine_tune_recipient(dag.vs[clade2], \n",
    "                                                              related_phyla[1])\n",
    "                phyla_relationship_flag = 1\n",
    "\n",
    "                if (recipient_taxon, donor_taxon) != related_phyla:\n",
    "                    descriptions[f'from {related_phyla[1]} to {related_phyla[0]}'].add('%s from %i to %i' % \\\n",
    "                                                                                       (candidate, donor_taxon, \n",
    "                                                                                        recipient_taxon))\n",
    "                else:\n",
    "                    descriptions[f'from {related_phyla[1]} to {related_phyla[0]}'].add(candidate)\n",
    "                continue\n",
    "\n",
    "            #\n",
    "            # if clade1 within clade2 ancestors it means that clade2 is nested within clade1\n",
    "            #   evidence of transfer from clade1 -> clade2\n",
    "            clade2_ancestors = udag.vs[clade2].get_shortest_paths(udag.vs[0])[0][1:]\n",
    "            if clade1 in clade2_ancestors:\n",
    "#                 support = dag.vs[clade1].in_edges()[0]['support']\n",
    "#                 print('%i is nested within %i             (%s)' % (related_phyla[1], related_phyla[0], candidate), support)\n",
    "                recipient_taxon         = fine_tune_recipient(dag.vs[clade2], \n",
    "                                                              related_phyla[1])\n",
    "                donor_taxon             = fine_tune_recipient(dag.vs[clade1], \n",
    "                                                              related_phyla[0])\n",
    "                phyla_relationship_flag = 1\n",
    "\n",
    "                if (donor_taxon, recipient_taxon) !=  related_phyla:\n",
    "                    descriptions[f'from {related_phyla[0]} to {related_phyla[1]}'].add('%s from %i to %i' % \\\n",
    "                                                                                       (candidate, donor_taxon, \n",
    "                                                                                        recipient_taxon))\n",
    "                else:\n",
    "                    descriptions[f'from {related_phyla[0]} to {related_phyla[1]}'].add(candidate)\n",
    "                continue\n",
    "\n",
    "            #\n",
    "            # if clade1 and clade2 are have the same parent noe it means they are sisters\n",
    "            #   evidence of hgt, but information about directionality\n",
    "            if clade1_ancestors[0] == clade2_ancestors[0]:\n",
    "#                 if dag.vs[clade1].predecessors()[0].index:\n",
    "#                     support = dag.vs[clade1].predecessors()[0].in_edges()[0]['support']\n",
    "#                     print('%i and %i are sisters              (%s)' % (related_phyla[0], related_phyla[1], candidate), support)\n",
    "                descriptions['sisters'].add(candidate)\n",
    "#                 else:\n",
    "#                     print('%i and %i are sisters              (%s)' % (related_phyla[0], related_phyla[1], candidate), 'root')\n",
    "                phyla_relationship_flag = 1\n",
    "                continue\n",
    "\n",
    "    #     with cd('candidates/'):\n",
    "    #         with open('%s-cluster%s.figTree' % (group_id, cluster_num), 'w') as out:\n",
    "    #             out.write(visualize_reconstruct_candidate(tmp_tree))\n",
    "\n",
    "        #\n",
    "        # if the flag still is False, we couldn't identify an relationship between phyla\n",
    "        if not phyla_relationship_flag:\n",
    "            descriptions['no clear relationship between phyla'].add(candidate)\n",
    "\n",
    "    return(descriptions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1090, 1117)\n",
      "(1090, 1224)\n",
      "(1090, 200795)\n",
      "(1090, 976)\n",
      "(1090, 1134404)\n",
      "(1090, 1798710)\n",
      "(1117, 1224)\n",
      "(1117, 200795)\n",
      "(1117, 976)\n",
      "(1117, 1134404)\n",
      "(1117, 1798710)\n",
      "(1224, 200795)\n",
      "(1224, 976)\n",
      "(1224, 1134404)\n",
      "(1224, 1798710)\n",
      "(200795, 976)\n",
      "(200795, 1134404)\n",
      "(200795, 1798710)\n",
      "(976, 1134404)\n",
      "(976, 1798710)\n",
      "(1134404, 1798710)\n"
     ]
    }
   ],
   "source": [
    "descriptions = {}\n",
    "for phylum_pair in itertools.combinations(sampled_phyla, 2):\n",
    "# for phylum_pair in [(200795, 1117)]:\n",
    "#     if 1224 in phylum_pair:\n",
    "#         continue\n",
    "    \n",
    "    print(phylum_pair)\n",
    "    descriptions[phylum_pair] = assess_transfers_between_phyla(phylum_pair)\n",
    "    \n",
    "# descriptions = assess_transfers_between_phyla((200795, 1117))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1090-1117\n",
      "   from 1090 to 1117: 1\n",
      "   from 1117 to 1090: 6\n",
      "   sisters: 5\n",
      "   no clear relationship between phyla: 11\n",
      "\n",
      "1090-200795\n",
      "   from 1090 to 200795: 1\n",
      "   from 200795 to 1090: 2\n",
      "   sisters: 5\n",
      "   no clear relationship between phyla: 4\n",
      "\n",
      "1090-976\n",
      "   from 1090 to 976: 0\n",
      "   from 976 to 1090: 171\n",
      "   sisters: 48\n",
      "   no clear relationship between phyla: 78\n",
      "\n",
      "1090-1134404\n",
      "   from 1090 to 1134404: 0\n",
      "   from 1134404 to 1090: 0\n",
      "   sisters: 0\n",
      "   no clear relationship between phyla: 0\n",
      "\n",
      "1090-1798710\n",
      "   from 1090 to 1798710: 0\n",
      "   from 1798710 to 1090: 0\n",
      "   sisters: 0\n",
      "   no clear relationship between phyla: 0\n",
      "\n",
      "1117-200795\n",
      "   from 1117 to 200795: 24\n",
      "   from 200795 to 1117: 18\n",
      "   sisters: 14\n",
      "   no clear relationship between phyla: 21\n",
      "\n",
      "1117-976\n",
      "   from 1117 to 976: 10\n",
      "   from 976 to 1117: 45\n",
      "   sisters: 17\n",
      "   no clear relationship between phyla: 30\n",
      "\n",
      "1117-1134404\n",
      "   from 1117 to 1134404: 0\n",
      "   from 1134404 to 1117: 0\n",
      "   sisters: 0\n",
      "   no clear relationship between phyla: 0\n",
      "\n",
      "1117-1798710\n",
      "   from 1117 to 1798710: 0\n",
      "   from 1798710 to 1117: 0\n",
      "   sisters: 0\n",
      "   no clear relationship between phyla: 0\n",
      "\n",
      "200795-976\n",
      "   from 200795 to 976: 0\n",
      "   from 976 to 200795: 5\n",
      "   sisters: 3\n",
      "   no clear relationship between phyla: 9\n",
      "\n",
      "200795-1134404\n",
      "   from 200795 to 1134404: 0\n",
      "   from 1134404 to 200795: 0\n",
      "   sisters: 0\n",
      "   no clear relationship between phyla: 0\n",
      "\n",
      "200795-1798710\n",
      "   from 200795 to 1798710: 0\n",
      "   from 1798710 to 200795: 0\n",
      "   sisters: 0\n",
      "   no clear relationship between phyla: 0\n",
      "\n",
      "976-1134404\n",
      "   from 976 to 1134404: 3\n",
      "   from 1134404 to 976: 0\n",
      "   sisters: 0\n",
      "   no clear relationship between phyla: 0\n",
      "\n",
      "976-1798710\n",
      "   from 976 to 1798710: 0\n",
      "   from 1798710 to 976: 0\n",
      "   sisters: 0\n",
      "   no clear relationship between phyla: 0\n",
      "\n",
      "1134404-1798710\n",
      "   from 1134404 to 1798710: 0\n",
      "   from 1798710 to 1134404: 0\n",
      "   sisters: 0\n",
      "   no clear relationship between phyla: 0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "candidates_to_improve = set()\n",
    "for pair, description in descriptions.items():\n",
    "    if 1224 in pair:\n",
    "        continue\n",
    "        \n",
    "    print('%i-%i' % pair)\n",
    "    for key, value in description.items():\n",
    "        if re.match('from \\d+ to \\d+$', key):\n",
    "            for candidate in value:\n",
    "                candidates_to_improve.add(candidate.split()[0])\n",
    "            \n",
    "        print('  ', key+':', len(value))\n",
    "    \n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {},
   "outputs": [],
   "source": [
    "for candidate in candidates_to_improve:\n",
    "    group_id, cluster_num = candidate.split('#')\n",
    "    \n",
    "    if not os.path.isfile(f'candidates/alignments/{group_id}-cluster{cluster_num}.aln'):\n",
    "        print(candidate)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('candidates/candidates_to_run_iqtree', 'w') as out:\n",
    "    \n",
    "    for candidate in candidates_to_improve:\n",
    "        group_id, cluster_num = candidate.split('#')\n",
    "        out.write(f'{group_id}-cluster{cluster_num}\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "COG4279#4\n",
      "32RWS#1\n",
      "COG5654#0 from 1224 to 33071\n",
      "COG1932#1\n",
      "COG5622#0\n",
      "COG4329#1\n",
      "COG4453#0\n",
      "COG2939#5\n",
      "COG3626#1\n",
      "COG0229#2\n",
      "COG1393#0\n",
      "COG0483#2\n",
      "COG2059#11\n",
      "COG3950#5\n",
      "COG1322#6\n",
      "COG4603#5\n",
      "COG0783#0\n",
      "COG1875#0\n",
      "COG3607#3\n",
      "COG0006#0 from 1224 to 33071\n",
      "COG5350#0 from 28211 to 1117\n",
      "31YEN#1\n",
      "COG2906#0\n",
      "COG1791#1\n",
      "32XNF#0 from 1236 to 1117\n",
      "COG4778#0\n",
      "COG0857#1\n",
      "COG2957#2\n",
      "COG4392#0\n",
      "331H5#1\n",
      "COG0229#2 from 1224 to 33071\n",
      "COG0792#0\n",
      "COG0354#1 from 28211 to 1117\n",
      "COG3861#0 from 1224 to 1416614\n",
      "COG4325#2\n",
      "32VVI#0\n",
      "COG5463#0\n",
      "COG0031#0 from 1224 to 1416614\n",
      "COG4633#3\n",
      "COG2243#3\n",
      "COG3454#4\n",
      "COG4328#3\n",
      "32TNJ#0 from 28211 to 1117\n",
      "COG0557#1\n",
      "COG1115#0\n",
      "32YVY#2\n",
      "COG0303#3\n",
      "COG4338#0\n",
      "COG1392#0\n",
      "COG0008#0\n",
      "COG0511#3\n",
      "COG4222#0\n",
      "COG2371#1 from 28211 to 1117\n",
      "COG1260#1\n",
      "COG5309#5\n",
      "COG0727#0\n",
      "COG0361#2\n",
      "COG4222#0 from 1224 to 33072\n",
      "COG0144#0\n",
      "COG3379#3\n",
      "COG1117#4\n",
      "COG1262#6\n",
      "COG1204#5\n",
      "COG0317#9\n",
      "COG3591#1\n",
      "COG3450#0\n",
      "COG0540#1\n",
      "COG0002#0\n",
      "COG3251#0\n",
      "COG3614#0\n",
      "COG1682#0\n",
      "COG3636#3\n",
      "COG3657#2\n",
      "COG5554#1\n",
      "COG5662#0 from 1224 to 33072\n",
      "COG3576#5\n",
      "COG0166#1\n",
      "COG0402#4\n",
      "COG3665#0\n",
      "COG1505#0 from 1224 to 1416614\n",
      "COG1959#1 from 1224 to 33071\n",
      "COG2831#3\n",
      "COG3227#3\n",
      "COG1112#7\n",
      "COG0242#2 from 1224 to 33072\n",
      "32ZHQ#0\n",
      "COG0560#2\n",
      "COG1741#1\n",
      "COG4456#3\n",
      "COG2372#1 from 1224 to 33071\n",
      "COG0397#5\n",
      "COG4222#4\n",
      "COG1573#2\n",
      "COG2162#3\n",
      "COG3807#0\n",
      "COG3665#1\n",
      "COG0401#0\n",
      "332XN#0\n",
      "COG1988#0\n",
      "COG0474#3\n",
      "COG2146#3\n",
      "COG0272#7\n",
      "COG4938#2\n",
      "COG0418#0\n",
      "COG0220#1\n",
      "COG1282#0\n",
      "3302Q#0\n",
      "COG3703#1\n",
      "COG0535#1\n",
      "COG2235#3\n",
      "COG0024#2 from 1224 to 33072\n",
      "COG0630#3\n",
      "COG4823#3\n",
      "COG2886#1\n",
      "COG2070#1\n",
      "COG3959#0\n",
      "2ZB6Z#2\n",
      "2ZCFZ#0\n",
      "COG0514#3\n",
      "COG4445#0\n",
      "COG0725#0 from 1224 to 1416614\n",
      "COG3803#0\n",
      "COG0398#1\n",
      "COG1294#0\n",
      "COG1805#1\n",
      "COG2372#1\n",
      "COG1814#5\n",
      "COG5267#2\n",
      "COG0636#7\n",
      "COG4325#1\n",
      "COG1528#7\n",
      "COG0129#8\n",
      "COG1155#1\n",
      "COG0350#2\n",
      "COG3524#2\n",
      "COG2193#3\n",
      "COG3558#3\n",
      "COG4683#4\n",
      "COG0465#13 from 28211 to 1117\n",
      "COG1918#0\n",
      "COG5266#4\n",
      "COG4451#2 from 1236 to 1117\n",
      "COG5450#3 from 80840 to 1117\n",
      "COG3932#1\n",
      "COG1598#3\n",
      "COG4420#6\n",
      "COG1640#0\n",
      "COG1285#2\n",
      "COG4551#0\n",
      "COG2932#0\n",
      "COG1742#2\n",
      "COG0719#4\n",
      "2Z7NV#1\n",
      "COG3895#0\n",
      "COG0591#6\n",
      "COG4232#1\n",
      "COG0450#0\n",
      "COG3865#1\n",
      "COG0350#2 from 1224 to 33071\n",
      "COG0678#0\n",
      "COG2032#3\n",
      "32ZWX#0\n",
      "3309S#1\n",
      "COG3350#6\n",
      "COG0529#1\n",
      "30DUI#0\n",
      "COG0441#6\n",
      "COG4576#4\n",
      "COG1791#2\n",
      "COG1898#2\n",
      "COG0287#7\n",
      "COG0499#5\n",
      "COG2823#0 from 1224 to 33071\n",
      "COG5626#1\n",
      "COG4319#13\n",
      "COG0552#1\n",
      "COG2214#2 from 28211 to 1117\n",
      "COG4875#4\n",
      "COG1317#2\n",
      "COG4249#5\n",
      "COG1054#0\n",
      "COG4683#4 from 1224 to 1416614\n",
      "COG0006#4\n",
      "COG1340#3\n",
      "COG3837#2\n",
      "COG0545#3\n",
      "COG4678#1\n",
      "COG0068#3\n",
      "2ZPQC#0\n",
      "COG1251#4\n",
      "COG0757#0\n",
      "COG0252#3\n",
      "COG1222#1\n",
      "COG2044#3\n",
      "COG0306#2\n",
      "COG5553#1\n",
      "COG1385#6\n",
      "COG0510#0\n",
      "32RYR#0\n",
      "COG1783#3\n",
      "COG2020#6\n",
      "COG0689#1\n",
      "COG3439#1\n",
      "COG3551#2\n",
      "COG1776#5\n",
      "COG1881#1\n",
      "COG2761#1\n",
      "COG0847#2\n",
      "COG0428#2\n",
      "COG1237#0\n",
      "COG3980#3\n",
      "COG3965#0\n",
      "COG4531#2\n",
      "COG3971#0\n",
      "COG1496#1\n",
      "COG0668#2\n",
      "COG3038#1\n",
      "COG2241#2\n",
      "COG5184#1\n",
      "COG5304#0\n",
      "COG3306#0\n",
      "COG3759#1\n",
      "COG3624#0\n",
      "COG5340#4\n",
      "COG1348#3\n",
      "2Z7N9#1\n",
      "COG0431#3\n",
      "COG3625#1\n",
      "COG0569#1\n",
      "COG0754#3\n",
      "COG3542#1\n",
      "2Z9VT#0\n",
      "COG1283#1\n",
      "COG1305#4\n",
      "COG3636#7\n",
      "COG4679#1\n",
      "COG0628#9 from 1224 to 33071\n",
      "COG2433#1\n",
      "330QS#1\n",
      "COG0409#2\n",
      "COG1916#5\n",
      "2Z8UG#1\n",
      "COG4113#2\n",
      "COG1218#4\n",
      "COG1598#1\n",
      "COG2002#8\n",
      "COG0246#1\n",
      "COG3161#1 from 1236 to 1117\n",
      "COG0240#1\n",
      "2ZAAA#2\n",
      "COG2336#5\n",
      "COG1271#0\n",
      "COG1076#0\n",
      "COG2040#1\n",
      "COG3021#2\n",
      "COG3954#0\n",
      "2Z8XI#0 from 1236 to 1117\n",
      "COG4126#3\n",
      "COG4798#3\n",
      "COG4683#0\n",
      "COG2516#2\n",
      "COG1269#5\n",
      "COG3638#4\n",
      "COG5625#0\n",
      "COG3950#5 from 1224 to 1416614\n",
      "COG0006#0\n",
      "COG3214#1\n",
      "2Z9FC#0\n",
      "32BAX#1 from 356 to 1117\n",
      "COG0278#0\n",
      "COG3041#0\n",
      "COG4714#2\n",
      "COG0023#0 from 1224 to 33071\n",
      "COG1858#3\n",
      "COG4456#3 from 1224 to 33072\n",
      "COG0579#2\n",
      "2Z7Q6#2\n",
      "COG0668#5\n",
      "COG1926#0\n",
      "COG3729#1\n",
      "2Z87F#2 from 28221 to 1117\n",
      "COG0393#2\n",
      "COG0510#0 from 445 to 1117\n",
      "COG1657#2\n",
      "COG3271#7\n",
      "COG0396#3 from 1224 to 33071\n",
      "COG0748#5\n",
      "32UZ7#0\n",
      "COG3510#0\n",
      "COG3239#3\n",
      "COG4076#4\n",
      "COG3576#4\n",
      "COG3571#3\n",
      "COG0302#4\n",
      "COG1230#0\n",
      "COG1626#1\n",
      "COG3431#1\n",
      "COG0374#1\n",
      "COG2886#6 from 1046 to 1117\n",
      "COG2327#5\n",
      "COG3019#0\n",
      "COG2877#1\n",
      "COG3608#4\n",
      "COG1515#3\n",
      "COG4102#0\n",
      "COG3221#3\n",
      "COG3614#1\n",
      "COG2333#3\n",
      "COG3615#0\n",
      "COG3329#3\n",
      "COG0512#1\n",
      "33AHD#0\n",
      "COG3695#1\n",
      "COG2603#1\n",
      "COG2370#0\n",
      "COG1721#1\n",
      "COG2082#2\n",
      "COG1449#2\n",
      "COG1820#2\n",
      "COG2828#1\n",
      "COG0145#0\n",
      "COG1305#3\n",
      "COG2886#6\n",
      "COG0038#8\n",
      "COG3585#0\n",
      "COG2944#8\n",
      "COG4097#2\n",
      "COG2520#0\n",
      "COG2193#2\n",
      "COG3624#1\n",
      "COG0119#7\n",
      "COG3268#3\n",
      "COG2929#2\n",
      "COG2223#4\n",
      "326SN#0\n",
      "COG0507#9\n",
      "COG1850#3\n",
      "COG4176#2\n",
      "COG1232#2\n",
      "COG2391#5\n",
      "COG3667#4\n",
      "32SSN#0\n",
      "COG0161#1\n",
      "COG3193#1\n",
      "COG3435#0\n",
      "COG1067#1\n",
      "COG0146#2\n",
      "COG2336#4\n",
      "COG1742#2 from 1224 to 33072\n",
      "COG3304#2\n",
      "COG1433#1\n",
      "COG0003#1\n",
      "2Z854#0\n",
      "COG4337#1\n",
      "COG3515#0\n",
      "COG3805#0\n",
      "COG4103#0\n",
      "COG4735#0\n",
      "COG3239#2\n",
      "COG4930#0\n",
      "COG3219#0\n",
      "32WB4#0 from 28211 to 1117\n",
      "COG0314#3\n",
      "COG2336#2\n",
      "31ARV#1\n",
      "COG2113#2\n",
      "COG2144#1\n",
      "COG2979#0\n",
      "COG4456#5\n",
      "COG0633#3\n",
      "32VXZ#3\n",
      "2Z8T4#1 from 135613 to 1117\n",
      "COG3510#3\n",
      "COG0701#5\n",
      "COG3220#4\n",
      "COG0695#4\n",
      "COG4456#3 from 1224 to 1416614\n",
      "COG1062#3\n",
      "COG5606#4\n",
      "COG2113#0\n",
      "COG3225#3\n",
      "COG0373#0\n",
      "32Z2Q#2\n",
      "2Z7N0#0\n",
      "COG2821#1 from 1236 to 1117\n",
      "COG1330#1\n",
      "COG2718#0\n",
      "COG5635#2\n",
      "COG4632#5\n",
      "COG1842#8\n",
      "COG3608#5 from 1224 to 33071\n",
      "COG0401#1\n",
      "COG0755#0\n",
      "COG4373#1\n",
      "COG0382#3\n",
      "COG3788#0\n",
      "COG3793#0\n",
      "COG5278#2\n",
      "COG3454#0\n",
      "COG3608#0\n",
      "COG2020#5\n",
      "COG2855#7\n",
      "COG2345#4\n",
      "COG4278#4\n",
      "COG5416#6\n",
      "COG0680#5\n",
      "COG0723#1\n",
      "COG4372#2\n",
      "COG0298#2\n",
      "COG1994#5\n",
      "COG4244#4\n",
      "COG4665#1\n",
      "COG1226#3\n",
      "COG3950#2 from 28221 to 1117\n",
      "COG1748#0\n",
      "COG3673#3\n",
      "COG3898#0\n",
      "COG0627#2 from 1224 to 33071\n",
      "COG2312#1\n",
      "COG3961#3\n",
      "COG0339#0\n",
      "COG0121#2\n",
      "COG3743#0\n",
      "COG2608#3\n",
      "COG0501#1\n",
      "COG4301#1\n",
      "COG3288#2\n",
      "COG3619#1\n",
      "COG0493#2\n",
      "COG3247#2\n",
      "COG3587#2\n",
      "COG0676#0\n",
      "32YHH#1\n",
      "COG3311#1\n",
      "COG4634#4\n",
      "COG2130#4\n",
      "COG2930#0\n",
      "COG5480#0\n",
      "COG1045#5\n",
      "COG0678#2\n",
      "COG2421#3\n",
      "COG1449#2 from 1236 to 1117\n",
      "COG2099#1\n",
      "COG0314#2\n",
      "COG1390#4\n",
      "COG3562#1\n",
      "COG3431#2\n",
      "COG4633#3 from 1224 to 1416614\n",
      "COG1959#1\n",
      "COG2194#2\n",
      "COG0247#6\n",
      "COG0573#0\n",
      "30CE8#0\n",
      "COG0315#3\n",
      "COG2049#1\n",
      "COG1505#0 from 1224 to 33072\n",
      "COG4337#3\n",
      "COG2193#5\n",
      "COG5304#2\n",
      "COG0694#2\n",
      "COG4671#0\n",
      "COG0480#6\n",
      "COG2988#2\n",
      "COG0864#1\n",
      "COG4067#0\n",
      "COG2764#1\n",
      "COG0430#1\n",
      "COG0380#0\n",
      "COG4680#0\n",
      "COG1682#10\n",
      "COG3546#2\n",
      "COG0260#0\n",
      "2Z7ZS#0\n",
      "COG4113#2 from 1224 to 1416614\n",
      "COG0119#3\n",
      "333DW#2\n",
      "COG1394#1\n",
      "COG0165#0\n",
      "COG0298#0\n",
      "COG3719#2\n",
      "COG0229#0\n",
      "COG0514#2\n",
      "COG5611#2\n",
      "COG3769#0\n",
      "COG0719#4 from 1224 to 33071\n",
      "COG4803#1\n",
      "COG1177#5\n",
      "COG4716#1\n",
      "COG3744#4\n",
      "COG1723#3\n",
      "COG3663#1\n",
      "COG0798#0\n",
      "COG3837#2 from 1224 to 33072\n",
      "COG2461#0\n",
      "COG0627#2\n",
      "COG4861#5\n",
      "COG0786#4\n",
      "COG0001#0\n",
      "COG3379#0\n",
      "COG4113#2 from 1224 to 33072\n",
      "COG2770#1\n",
      "COG0863#4 from 1224 to 1416614\n",
      "330AH#2\n",
      "COG0153#0\n",
      "COG3484#0\n",
      "COG1876#4\n",
      "COG3839#0 from 28211 to 1117\n",
      "COG3751#2\n",
      "COG2823#0\n",
      "COG0725#0\n",
      "COG0695#2\n",
      "COG2954#1\n",
      "COG0439#3\n",
      "COG0521#2\n",
      "COG0009#1\n",
      "COG1056#1\n",
      "31G5S#1\n",
      "2Z8IF#0\n",
      "COG2041#4\n",
      "COG0830#3\n",
      "COG3349#2\n",
      "COG3806#1\n",
      "COG4233#0 from 1224 to 33071\n",
      "2ZA6T#1\n",
      "COG0716#4\n",
      "COG3424#4\n",
      "COG0242#2\n",
      "COG5610#0\n",
      "33BC0#0\n",
      "COG4902#2\n",
      "32NM5#2\n",
      "32BAX#1\n",
      "COG1156#2\n",
      "COG0618#1\n",
      "COG0011#2\n",
      "COG3126#0\n",
      "COG0695#3\n",
      "COG2014#3\n",
      "33E3H#0 from 1236 to 1117\n",
      "COG5654#0\n",
      "COG5007#0\n",
      "COG4226#2\n",
      "COG0418#1\n",
      "COG4319#9\n",
      "32YT2#0\n",
      "2Z9EI#0\n",
      "COG0568#1\n",
      "COG5331#0\n",
      "COG2268#4 from 29 to 1117\n",
      "COG0023#0\n",
      "COG3755#4\n",
      "COG1363#2\n",
      "COG5483#2\n",
      "COG3639#4\n",
      "COG3326#0\n",
      "32S87#1\n",
      "COG3447#0\n",
      "COG3812#0\n",
      "COG4467#0\n",
      "COG4941#1\n",
      "COG0563#5\n",
      "COG4113#6\n",
      "COG3861#0\n",
      "COG2242#2\n",
      "COG1171#1\n",
      "COG3805#3 from 204441 to 1117\n",
      "COG2251#3\n",
      "2Z7WD#3\n",
      "COG1752#3\n",
      "COG0220#1 from 1224 to 33072\n",
      "COG3188#1\n",
      "COG5018#1\n",
      "COG3174#2\n",
      "COG0759#5\n",
      "COG0242#0\n",
      "COG4233#0\n",
      "COG5031#4 from 29 to 1117\n",
      "COG1083#2\n",
      "COG2321#2\n",
      "COG3686#1\n",
      "COG0033#2\n",
      "32YQJ#0\n",
      "COG0155#5\n",
      "COG0372#0\n",
      "COG2886#3\n",
      "2Z7NV#3\n",
      "COG5526#1\n",
      "COG0191#3\n",
      "COG4552#0\n",
      "COG3128#2\n",
      "COG2046#3\n",
      "COG0854#2\n",
      "COG3609#6\n",
      "COG1359#2\n",
      "COG1505#0\n",
      "COG1433#3\n",
      "COG1914#3\n",
      "COG2994#0\n",
      "COG4529#1\n",
      "COG1957#3\n",
      "COG2224#4 from 28211 to 1117\n",
      "COG3135#5\n",
      "COG0308#0 from 1224 to 33071\n",
      "COG5634#4\n",
      "COG5554#0 from 28211 to 1117\n",
      "COG0069#0\n",
      "COG3287#0\n",
      "COG0031#0\n",
      "COG4603#6\n",
      "COG4320#2\n",
      "COG0720#0\n",
      "COG1149#3\n",
      "COG3719#0\n",
      "COG1830#4\n",
      "COG1544#0\n",
      "COG1640#0 from 1224 to 33071\n",
      "COG0709#3\n",
      "COG4453#1\n",
      "COG3585#2\n",
      "COG1720#2\n",
      "COG2214#2\n",
      "COG3675#4\n",
      "333DW#2 from 28221 to 1117\n",
      "COG3491#3\n",
      "COG0722#1\n",
      "COG5493#2\n",
      "COG0398#0\n",
      "COG2220#4\n",
      "COG2193#6\n",
      "COG0467#5\n",
      "COG2002#3\n",
      "COG3657#0\n",
      "COG1643#2\n",
      "COG0057#4\n",
      "COG2391#2\n",
      "COG3514#1\n",
      "COG4076#0\n",
      "COG0655#3\n",
      "COG0831#2\n",
      "330AH#2 from 1224 to 1416614\n",
      "COG0033#0\n",
      "COG1230#2\n",
      "COG3900#0\n",
      "COG1671#2\n",
      "COG2906#3\n",
      "COG4935#5\n",
      "COG2233#4\n",
      "COG0182#1\n",
      "COG0141#3\n",
      "COG0684#1\n",
      "COG4322#1\n",
      "COG3754#1\n",
      "COG0308#5\n",
      "COG0645#5\n",
      "COG0769#1\n",
      "COG3751#3\n",
      "COG3950#2\n",
      "COG0288#0\n",
      "COG3563#1\n",
      "COG5499#3\n",
      "COG0024#2\n",
      "2Z8UT#0\n",
      "COG0226#3\n",
      "COG3280#1\n",
      "COG4968#1\n",
      "COG4227#2\n",
      "COG1984#2\n",
      "COG1402#3\n",
      "COG3812#1\n",
      "2Z9YC#0 from 29 to 1117\n",
      "COG1819#1\n",
      "COG1938#0\n",
      "COG4310#1 from 356 to 1117\n",
      "COG0628#9\n",
      "COG1178#0\n",
      "COG0757#3\n",
      "COG1234#0\n",
      "32EPT#0\n",
      "COG5316#3\n",
      "COG4245#5\n",
      "COG5361#5\n",
      "COG1446#2\n",
      "COG2509#3\n",
      "COG0863#4\n",
      "COG5563#2\n",
      "COG1524#5\n",
      "333NE#0\n",
      "COG3609#5\n",
      "COG1487#3\n",
      "COG4454#2\n",
      "COG0006#3\n",
      "COG2263#3\n",
      "COG3686#1 from 28211 to 1117\n",
      "COG5662#0\n"
     ]
    }
   ],
   "source": [
    "print('\\n'.join(descriptions[(1117,1224)]['from 1224 to 1117']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'from 200795 to 1117': {'COG0001#4',\n",
       "  'COG0049#2',\n",
       "  'COG0345#1',\n",
       "  'COG0354#0',\n",
       "  'COG0499#0',\n",
       "  'COG0554#3',\n",
       "  'COG0637#4',\n",
       "  'COG1075#1 from 32064 to 1117',\n",
       "  'COG1086#8',\n",
       "  'COG1177#0',\n",
       "  'COG1304#5',\n",
       "  'COG1336#1 from 32064 to 1117',\n",
       "  'COG1351#7 from 32061 to 1117',\n",
       "  'COG1554#0',\n",
       "  'COG1704#1',\n",
       "  'COG2896#1',\n",
       "  'COG3259#2',\n",
       "  'COG3379#4 from 32064 to 1117'},\n",
       " 'from 1117 to 200795': {'COG0115#2 from 1117 to 189775',\n",
       "  'COG0828#0',\n",
       "  'COG1517#2 from 1117 to 32061',\n",
       "  'COG1666#2',\n",
       "  'COG1894#0 from 1117 to 32064',\n",
       "  'COG1905#1 from 1117 to 133453',\n",
       "  'COG1905#1 from 1117 to 32064',\n",
       "  'COG1941#3 from 1117 to 120961',\n",
       "  'COG1941#3 from 1117 to 32064',\n",
       "  'COG2343#0 from 1117 to 104176',\n",
       "  'COG2343#0 from 1117 to 120961',\n",
       "  'COG2452#1',\n",
       "  'COG2981#4 from 1117 to 32064',\n",
       "  'COG4118#7 from 1117 to 32064',\n",
       "  'COG4467#0 from 1117 to 32064',\n",
       "  'COG4663#1',\n",
       "  'COG4978#0',\n",
       "  'COG4978#0 from 1117 to 363277',\n",
       "  'COG5000#0 from 1117 to 133453',\n",
       "  'COG5000#0 from 1117 to 2057',\n",
       "  'COG5000#0 from 1117 to 32064',\n",
       "  'COG5000#0 from 1117 to 65',\n",
       "  'COG5207#0 from 1117 to 363277',\n",
       "  'COG5635#3'},\n",
       " 'sisters': set(),\n",
       " 'no clear relationship between phyla': set()}"
      ]
     },
     "execution_count": 134,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "related_phyla = (200795, 1117)\n",
    "descriptions  = {f'from {related_phyla[0]} to {related_phyla[1]}': set(),\n",
    "                 f'from {related_phyla[1]} to {related_phyla[0]}': set(),\n",
    "                  'sisters':                                       set(),\n",
    "                  'no clear relationship between phyla':           set()}\n",
    "candidates    = set(significant_df.query('source==\"%i\" and target==\"%i\"' % related_phyla).group.values).union(\n",
    "                    significant_df.query('target==\"%i\" and source==\"%i\"' % related_phyla).group.values\n",
    "                )\n",
    "\n",
    "for candidate in candidates:\n",
    "\n",
    "    group_id, cluster_num = candidate.split('#')\n",
    "\n",
    "    with cd('candidates/trees/'):\n",
    "        #\n",
    "        # test if there is rooted version of the tree (mad output) \n",
    "        if not os.path.isfile('%s-cluster%s.fastTree.rooted' % (group_id, cluster_num)):\n",
    "            continue\n",
    "\n",
    "        #\n",
    "        # we need the original tree cause mad removes support values, so we just transfer root positions\n",
    "        try:\n",
    "            tmp_tree   = match_rooting(ete3.Tree('%s-cluster%s.fastTree.rooted' % (group_id, cluster_num)),\n",
    "                                       ete3.Tree('%s-cluster%s.fastTree'        % (group_id, cluster_num)))\n",
    "        except ete3.parser.newick.NewickError:\n",
    "            continue\n",
    "\n",
    "    #\n",
    "    # extract taxIDs to subsample taxonomy tables\n",
    "    taxids = {}\n",
    "    for leaf in tmp_tree.get_leaf_names():\n",
    "        #\n",
    "        # leaf names are composed by <taxid>.<locus_tag>\n",
    "        #   ps: locus_tag may also have \".\" within it\n",
    "        tmp_taxid = int(leaf.split('.')[0])\n",
    "\n",
    "        if tmp_taxid not in taxids:\n",
    "            taxids[tmp_taxid] = []\n",
    "        taxids[tmp_taxid].append(leaf)\n",
    "\n",
    "    #\n",
    "    # select taxIDs from each assessed phylum...\n",
    "    phylum1_taxonomy = eggNOG_taxonomy.loc[taxids].query('phylum==@related_phyla[0]')\n",
    "    phylum2_taxonomy = eggNOG_taxonomy.loc[taxids].query('phylum==@related_phyla[1]')\n",
    "\n",
    "    #\n",
    "    # ... and their respect leaves\n",
    "    phylum1_leaves = set()\n",
    "    for taxid in phylum1_taxonomy.index:\n",
    "        phylum1_leaves.update(taxids[taxid])\n",
    "\n",
    "    phylum2_leaves = set()\n",
    "    for taxid in phylum2_taxonomy.index:\n",
    "        phylum2_leaves.update(taxids[taxid])\n",
    "\n",
    "    all_leaves = phylum1_leaves.union(phylum2_leaves)\n",
    "\n",
    "    #\n",
    "    # as tree traversing through ete3 isn't very efficient, and doesn't scalate very well,\n",
    "    #   we create an iGraph dag for more effient traversing\n",
    "    dag  = tree_to_dag(tmp_tree)          # we need a directed version...\n",
    "    udag = dag.as_undirected(mode='each') # ... and an undirected one for different processes\n",
    "\n",
    "    #\n",
    "    # placeholders where we will add monophyletic clades for each phylum\n",
    "    phylum1_clades = set() \n",
    "    phylum2_clades = set()\n",
    "\n",
    "    ignored_nodes  = [] # descendants of monophyletic nodes should be ignored once their\n",
    "                        #   ancestors have been stored.\n",
    "\n",
    "    #\n",
    "    # traverse through internal nodes\n",
    "    for node in dag.vs:\n",
    "        if node.index in ignored_nodes:\n",
    "            continue\n",
    "\n",
    "        node_leaves = get_leaf_names(node)\n",
    "\n",
    "        if all_leaves.isdisjoint(node_leaves):\n",
    "            ignored_nodes.extend(get_descendant_indices(node, leaves=True))\n",
    "\n",
    "        #\n",
    "        # if there aren't leaves from other phyla within this node it is monophyletic\n",
    "        if phylum1_leaves.intersection(node_leaves) and phylum1_leaves.issuperset(node_leaves):\n",
    "            phylum1_clades.add(node.index)\n",
    "            ignored_nodes.extend(get_descendant_indices(node, leaves=True))\n",
    "\n",
    "        elif phylum2_leaves.intersection(node_leaves) and phylum2_leaves.issuperset(node_leaves):\n",
    "            phylum2_clades.add(node.index)\n",
    "            ignored_nodes.extend(get_descendant_indices(node, leaves=True))\n",
    "\n",
    "    #\n",
    "    # now we add some flexibility to the monophyly of nodes for three reasons:\n",
    "    #   1) donor nodes within gene trees will never be monophyletic since the recipient\n",
    "    #      must be nested within it.\n",
    "    #   2) if there are other transfers from the donor and/or recipient phyla to a 3rd one\n",
    "    #      we still want to capture it.\n",
    "    #   3) good ol' phylogenetic uncertainty\n",
    "    phylum1_clades = merge_polyphyletic_clades(phylum1_clades, udag)\n",
    "    phylum2_clades = merge_polyphyletic_clades(phylum2_clades, udag)\n",
    "\n",
    "    #\n",
    "    # flag if we can identify relations between phyla:\n",
    "    #   1 nested within 2\n",
    "    #   2 nested within 1\n",
    "    #   1 and 2 are sisters\n",
    "    #\n",
    "    # if no relation is identified, leave as false\n",
    "    phyla_relationship_flag = 0\n",
    "\n",
    "    for clade1, clade2 in itertools.product(phylum1_clades, phylum2_clades):\n",
    "\n",
    "        clade1_ancestors = udag.vs[clade1].get_shortest_paths(udag.vs[0])[0][1:]\n",
    "        if clade2 in clade1_ancestors:\n",
    "            recipient_taxon         = fine_tune_recipient(dag.vs[clade1], \n",
    "                                                          related_phyla[0])\n",
    "            donor_taxon             = fine_tune_recipient(dag.vs[clade2], \n",
    "                                                          related_phyla[1])\n",
    "            phyla_relationship_flag = 1\n",
    "            \n",
    "            if (recipient_taxon, donor_taxon) != related_phyla:\n",
    "                descriptions[f'from {related_phyla[1]} to {related_phyla[0]}'].add('%s from %i to %i' % \\\n",
    "                                                                                   (candidate, donor_taxon, \n",
    "                                                                                    recipient_taxon))\n",
    "            else:\n",
    "                descriptions[f'from {related_phyla[1]} to {related_phyla[0]}'].add(candidate)\n",
    "            continue\n",
    "\n",
    "        clade2_ancestors = udag.vs[clade2].get_shortest_paths(udag.vs[0])[0][1:]\n",
    "        if clade1 in clade2_ancestors:\n",
    "            recipient_taxon         = fine_tune_recipient(dag.vs[clade2], \n",
    "                                                          related_phyla[1])\n",
    "            donor_taxon             = fine_tune_recipient(dag.vs[clade1], \n",
    "                                                          related_phyla[0])\n",
    "            phyla_relationship_flag = 1\n",
    "            \n",
    "            if (donor_taxon, recipient_taxon) !=  related_phyla:\n",
    "                descriptions[f'from {related_phyla[0]} to {related_phyla[1]}'].add('%s from %i to %i' % \\\n",
    "                                                                                   (candidate, donor_taxon, \n",
    "                                                                                    recipient_taxon))\n",
    "            else:\n",
    "                descriptions[f'from {related_phyla[0]} to {related_phyla[1]}'].add(candidate)\n",
    "            continue\n",
    "\n",
    "descriptions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "32064.0"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fine_tune_recipient(dag.vs[clade1], related_phyla[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('COG4118#7', 1117.0, None)"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "candidate, donor_taxon, recipient_taxon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(94, 2)"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clade1, clade2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fine_tune_recipient(node, phylum):\n",
    "    node_taxids   = [int(leaf.split('.')[0]) for leaf in get_leaf_names(node)]\n",
    "    node_taxonomy = eggNOG_taxonomy.loc[node_taxids, \n",
    "                                             ['phylum', 'class', 'order', \n",
    "                                              'family', 'genus', 'species']\n",
    "                                            ].query(f'phylum=={phylum}')\n",
    "\n",
    "    most_specific_taxon = None\n",
    "    for rank, column in node_taxonomy.iteritems():\n",
    "        unique_taxa = column.unique()\n",
    "\n",
    "        if pd.notna(unique_taxa).all() and len(unique_taxa) == 1:\n",
    "            most_specific_taxon = unique_taxa[0]\n",
    "        else:\n",
    "            break\n",
    "    \n",
    "    return(most_specific_taxon)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "igraph.Vertex(<igraph.Graph object at 0x2aab6b271250>, 94, {'name': 'node_94', 'is_leaf': False})"
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dag.vs[clade1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [],
   "source": [
    "node_taxids   = [int(leaf.split('.')[0]) for leaf in get_leaf_names(dag.vs[clade1])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1521187, 383372, 326427, 324602, 357808]"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "node_taxids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>class</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>phylum</th>\n",
       "      <th>order</th>\n",
       "      <th>species</th>\n",
       "      <th>superkingdom</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1521187</th>\n",
       "      <td>32061.0</td>\n",
       "      <td>1106.0</td>\n",
       "      <td>1107.0</td>\n",
       "      <td>200795.0</td>\n",
       "      <td>32064.0</td>\n",
       "      <td>1521187.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>383372</th>\n",
       "      <td>32061.0</td>\n",
       "      <td>1508635.0</td>\n",
       "      <td>120961.0</td>\n",
       "      <td>200795.0</td>\n",
       "      <td>32064.0</td>\n",
       "      <td>120962.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>326427</th>\n",
       "      <td>32061.0</td>\n",
       "      <td>1106.0</td>\n",
       "      <td>1107.0</td>\n",
       "      <td>200795.0</td>\n",
       "      <td>32064.0</td>\n",
       "      <td>152260.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>324602</th>\n",
       "      <td>32061.0</td>\n",
       "      <td>1106.0</td>\n",
       "      <td>1107.0</td>\n",
       "      <td>200795.0</td>\n",
       "      <td>32064.0</td>\n",
       "      <td>1108.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357808</th>\n",
       "      <td>32061.0</td>\n",
       "      <td>1508635.0</td>\n",
       "      <td>120961.0</td>\n",
       "      <td>200795.0</td>\n",
       "      <td>32064.0</td>\n",
       "      <td>357808.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           class     family     genus    phylum    order    species  \\\n",
       "index                                                                 \n",
       "1521187  32061.0     1106.0    1107.0  200795.0  32064.0  1521187.0   \n",
       "383372   32061.0  1508635.0  120961.0  200795.0  32064.0   120962.0   \n",
       "326427   32061.0     1106.0    1107.0  200795.0  32064.0   152260.0   \n",
       "324602   32061.0     1106.0    1107.0  200795.0  32064.0     1108.0   \n",
       "357808   32061.0  1508635.0  120961.0  200795.0  32064.0   357808.0   \n",
       "\n",
       "         superkingdom  \n",
       "index                  \n",
       "1521187           2.0  \n",
       "383372            2.0  \n",
       "326427            2.0  \n",
       "324602            2.0  \n",
       "357808            2.0  "
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eggNOG_taxonomy.loc[node_taxids]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "32064.0"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fine_tune_recipient(dag.vs[clade1], 200795)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
