{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/nobackup1b/users/thiberio/eggNOG\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "import itertools\n",
    "import numpy as np\n",
    "from matplotlib import pyplot as plt\n",
    "import seaborn as sns\n",
    "import igraph as ig\n",
    "from IPython.display import HTML\n",
    "import multiprocessing\n",
    "import pickle as pkl\n",
    "import random\n",
    "from collections import Counter\n",
    "import os\n",
    "import subprocess\n",
    "from scipy.stats import mannwhitneyu\n",
    "import ete3\n",
    "from copy import deepcopy\n",
    "from math import ceil\n",
    "\n",
    "ncbi = ete3.NCBITaxa()\n",
    "%run assess_connections-functions.ipynb\n",
    "\n",
    "%cd ~/work/eggNOG/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "sampled_genomes = pd.read_csv('../kelsey/genomes.tab',\n",
    "                              sep='\\t',\n",
    "                              index_col=0)\n",
    "\n",
    "lineages = pd.DataFrame()\n",
    "for taxid in sampled_genomes.species_taxid.unique():\n",
    "    if pd.isna(taxid):\n",
    "        continue\n",
    "    lineages = lineages.append({tax_rank: tmp_taxid \n",
    "                                 for tmp_taxid, tax_rank in ncbi.get_rank(ncbi.get_lineage(taxid)).items()},\n",
    "                                ignore_index=True)\n",
    "lineages = lineages.reindex(columns=['class', 'family',  'genus', 'phylum',\n",
    "                                     'order', 'species', 'superkingdom']).copy()\n",
    "lineages = lineages.query('superkingdom == 2').copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "sampled_phyla = [phylum for phylum in lineages.phylum.unique().astype(int) if phylum > 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "working_groups  = pd.read_parquet('working_eggNOG_groups.parquet', engine='fastparquet')\n",
    "working_trees   = pd.read_parquet('working_eggNOG_trees.parquet' , engine='fastparquet')\n",
    "eggNOG_taxonomy = pd.read_parquet('eggNOG_taxonomy.parquet'      , engine='fastparquet')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('all_results.pkl', 'rb') as _:\n",
    "    phylum_connections = pkl.load(_)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "regular_connections     = []\n",
    "significant_connections = []\n",
    "\n",
    "for group_id, group_clusters in phylum_connections:\n",
    "    if not group_clusters:\n",
    "        continue\n",
    "    \n",
    "    for cluster_num, cluster_data in group_clusters.items():\n",
    "\n",
    "        for ref_phylum, phyla_dists in cluster_data.items():\n",
    "            \n",
    "            if not phyla_dists['df'].shape[0]:\n",
    "                continue\n",
    "\n",
    "            closest_phylum = phyla_dists['df'].iloc[0, 0]\n",
    "            if not closest_phylum in sampled_phyla:\n",
    "                continue\n",
    "\n",
    "            if phyla_dists['significant']:\n",
    "                significant_connections.append( (ref_phylum,\n",
    "                                                 phyla_dists['df'].iloc[0, 0],\n",
    "                                                 phyla_dists['df'].iloc[0, 1], \n",
    "                                                 '%s#%i' % (group_id, cluster_num)) )\n",
    "\n",
    "                for index, tmp_series in phyla_dists['df'].iloc[1:, :].iterrows():\n",
    "                    if tmp_series.phylum in sampled_phyla:\n",
    "                        regular_connections.append( (ref_phylum,\n",
    "                                                         tmp_series.phylum,\n",
    "                                                         tmp_series['median'], \n",
    "                                                         '%s#%i' % (group_id, cluster_num)) )\n",
    "\n",
    "            else:\n",
    "                for index, tmp_series in phyla_dists['df'].iterrows():\n",
    "                    if tmp_series.phylum in sampled_phyla:\n",
    "                        regular_connections.append( (ref_phylum,\n",
    "                                                         tmp_series.phylum,\n",
    "                                                         tmp_series['median'], \n",
    "                                                         '%s#%i' % (group_id, cluster_num)) )\n",
    "\n",
    "significant_df = pd.DataFrame(data=significant_connections, columns=['source', 'target', 'distance', 'group'])\n",
    "regular_df     = pd.DataFrame(data=regular_connections,     columns=['source', 'target', 'distance', 'group'])\n",
    "\n",
    "significant_df.dropna(how='any', inplace=True)\n",
    "regular_df.dropna(    how='any', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# iqtree\n",
    "#\n",
    "\n",
    "def assess_transfers_between_phyla(related_phyla, candidates):\n",
    "# related_phyla = (1090, 1117)\n",
    "\n",
    "    descriptions  = {f'from {related_phyla[0]} to {related_phyla[1]}': set(),\n",
    "                     f'from {related_phyla[1]} to {related_phyla[0]}': set(),\n",
    "                      'sisters':                                       set(),\n",
    "                      'no clear relationship between phyla':           set()}\n",
    "\n",
    "    for candidate in candidates:\n",
    "\n",
    "        group_id, cluster_num = candidate.split('#')\n",
    "\n",
    "#         if os.path.isfile('candidates/missing_iqtree/%s-cluster%s.treefile.rooted' % (group_id, \n",
    "#                                                                               cluster_num)):\n",
    "#             with cd('candidates/missing_iqtree/'):\n",
    "#                 #\n",
    "#                 # we need the original tree cause mad removes support values, so we just transfer root positions\n",
    "#                 try:\n",
    "#                     tmp_tree   = match_rooting(ete3.Tree('%s-cluster%s.treefile.rooted' % (group_id, \n",
    "#                                                                                            cluster_num),\n",
    "#                                                          format=1),\n",
    "#                                                ete3.Tree('%s-cluster%s.treefile'        % (group_id, \n",
    "#                                                                                            cluster_num),\n",
    "#                                                          format=1))\n",
    "#                 except ete3.parser.newick.NewickError:\n",
    "#                     continue\n",
    "        if os.path.isfile('candidates/iqtree/%s-cluster%s.treefile.rooted' % (group_id, \n",
    "                                                                              cluster_num)):\n",
    "            with cd('candidates/iqtree/'):\n",
    "                #\n",
    "                # we need the original tree cause mad removes support values, so we just transfer root positions\n",
    "                try:\n",
    "                    tmp_tree   = match_rooting(ete3.Tree('%s-cluster%s.treefile.rooted' % (group_id, \n",
    "                                                                                           cluster_num),\n",
    "                                                         format=1),\n",
    "                                               ete3.Tree('%s-cluster%s.treefile'        % (group_id, \n",
    "                                                                                           cluster_num),\n",
    "                                                         format=1))\n",
    "                except ete3.parser.newick.NewickError:\n",
    "                    continue\n",
    "        else:\n",
    "            print(candidate)\n",
    "            continue\n",
    "\n",
    "        #\n",
    "        # extract taxIDs to subsample taxonomy tables\n",
    "        taxids = {}\n",
    "        for leaf in tmp_tree.get_leaf_names():\n",
    "            #\n",
    "            # leaf names are composed by <taxid>.<locus_tag>\n",
    "            #   ps: locus_tag may also have \".\" within it\n",
    "            tmp_taxid = int(leaf.split('.')[0])\n",
    "\n",
    "            if tmp_taxid not in taxids:\n",
    "                taxids[tmp_taxid] = []\n",
    "            taxids[tmp_taxid].append(leaf)\n",
    "\n",
    "        #\n",
    "        # select taxIDs from each assessed phylum...\n",
    "        phylum1_taxonomy = eggNOG_taxonomy.loc[taxids].query('phylum==@related_phyla[0]')\n",
    "        phylum2_taxonomy = eggNOG_taxonomy.loc[taxids].query('phylum==@related_phyla[1]')\n",
    "\n",
    "        #\n",
    "        # ... and their respect leaves\n",
    "        phylum1_leaves = set()\n",
    "        for taxid in phylum1_taxonomy.index:\n",
    "            phylum1_leaves.update(taxids[taxid])\n",
    "\n",
    "        phylum2_leaves = set()\n",
    "        for taxid in phylum2_taxonomy.index:\n",
    "            phylum2_leaves.update(taxids[taxid])\n",
    "        \n",
    "        all_leaves = phylum1_leaves.union(phylum2_leaves)\n",
    "\n",
    "        #\n",
    "        # as tree traversing through ete3 isn't very efficient, and doesn't scalate very well,\n",
    "        #   we create an iGraph dag for more effient traversing\n",
    "        dag  = tree_to_dag_bb_and_alrt(tmp_tree)          # we need a directed version...\n",
    "        udag = dag.as_undirected(mode='each') # ... and an undirected one for different processes\n",
    "\n",
    "        #\n",
    "        # placeholders where we will add monophyletic clades for each phylum\n",
    "        phylum1_clades = set() \n",
    "        phylum2_clades = set()\n",
    "\n",
    "        ignored_nodes  = [] # descendants of monophyletic nodes should be ignored once their\n",
    "                            #   ancestors have been stored.\n",
    "\n",
    "        #\n",
    "        # traverse through internal nodes\n",
    "        for node in dag.vs:\n",
    "            if node.index in ignored_nodes:\n",
    "                continue\n",
    "\n",
    "            node_leaves = get_leaf_names(node)\n",
    "\n",
    "            if all_leaves.isdisjoint(node_leaves):\n",
    "                ignored_nodes.extend(get_descendant_indices(node, leaves=True))\n",
    "\n",
    "            #\n",
    "            # if there aren't leaves from other phyla within this node it is monophyletic\n",
    "            if phylum1_leaves.intersection(node_leaves) and phylum1_leaves.issuperset(node_leaves):\n",
    "                phylum1_clades.add(node.index)\n",
    "                ignored_nodes.extend(get_descendant_indices(node, leaves=True))\n",
    "\n",
    "            elif phylum2_leaves.intersection(node_leaves) and phylum2_leaves.issuperset(node_leaves):\n",
    "                phylum2_clades.add(node.index)\n",
    "                ignored_nodes.extend(get_descendant_indices(node, leaves=True))\n",
    "\n",
    "        #\n",
    "        # now we add some flexibility to the monophyly of nodes for three reasons:\n",
    "        #   1) donor nodes within gene trees will never be monophyletic since the recipient\n",
    "        #      must be nested within it.\n",
    "        #   2) if there are other transfers from the donor and/or recipient phyla to a 3rd one\n",
    "        #      we still want to capture it.\n",
    "        #   3) good ol' phylogenetic uncertainty\n",
    "        phylum1_clades = merge_polyphyletic_clades(phylum1_clades, udag)\n",
    "        phylum2_clades = merge_polyphyletic_clades(phylum2_clades, udag)\n",
    "\n",
    "        #\n",
    "        # flag if we can identify relations between phyla:\n",
    "        #   1 nested within 2\n",
    "        #   2 nested within 1\n",
    "        #   1 and 2 are sisters\n",
    "        #\n",
    "        # if no relation is identified, leave as false\n",
    "        phyla_relationship_flag = 0\n",
    "\n",
    "        for clade1, clade2 in itertools.product(phylum1_clades, phylum2_clades):\n",
    "\n",
    "            #\n",
    "            # capture ancestors of clade1 by querying nodes between itself and the root node\n",
    "            clade1_ancestors = udag.vs[clade1].get_shortest_paths(udag.vs[0])[0][1:]\n",
    "            #\n",
    "            # if clade2 within clade1 ancestors it means that clade1 is nested within clade2\n",
    "            #   evidence of transfer from clade2 -> clade1\n",
    "            if clade2 in clade1_ancestors:\n",
    "                recipient_taxon         = fine_tune_recipient(dag.vs[clade1], \n",
    "                                                              related_phyla[0])\n",
    "                donor_taxon             = fine_tune_recipient(dag.vs[clade2], \n",
    "                                                              related_phyla[1])\n",
    "                phyla_relationship_flag = 1\n",
    "                \n",
    "                if (recipient_taxon, donor_taxon) != related_phyla:\n",
    "                    descriptions[f'from {related_phyla[1]} to {related_phyla[0]}'].add('%s from %i to %i' % \\\n",
    "                                                                                       (candidate, donor_taxon, \n",
    "                                                                                        recipient_taxon))\n",
    "                else:\n",
    "                    descriptions[f'from {related_phyla[1]} to {related_phyla[0]}'].add(candidate)\n",
    "                continue\n",
    "\n",
    "            #\n",
    "            # if clade1 within clade2 ancestors it means that clade2 is nested within clade1\n",
    "            #   evidence of transfer from clade1 -> clade2\n",
    "            clade2_ancestors = udag.vs[clade2].get_shortest_paths(udag.vs[0])[0][1:]\n",
    "            if clade1 in clade2_ancestors:\n",
    "                recipient_taxon         = fine_tune_recipient(dag.vs[clade2], \n",
    "                                                              related_phyla[1])\n",
    "                donor_taxon             = fine_tune_recipient(dag.vs[clade1], \n",
    "                                                              related_phyla[0])\n",
    "                phyla_relationship_flag = 1\n",
    "\n",
    "                if (donor_taxon, recipient_taxon) !=  related_phyla:\n",
    "                    descriptions[f'from {related_phyla[0]} to {related_phyla[1]}'].add('%s from %i to %i' % \\\n",
    "                                                                                       (candidate, donor_taxon, \n",
    "                                                                                        recipient_taxon))\n",
    "                else:\n",
    "                    descriptions[f'from {related_phyla[0]} to {related_phyla[1]}'].add(candidate)\n",
    "\n",
    "                continue\n",
    "\n",
    "            #\n",
    "            # if clade1 and clade2 are have the same parent noe it means they are sisters\n",
    "            #   evidence of hgt, but information about directionality\n",
    "            if clade1_ancestors[0] == clade2_ancestors[0]:\n",
    "                descriptions['sisters'].add(candidate)\n",
    "#                 else:\n",
    "#                     print('%i and %i are sisters              (%s)' % (related_phyla[0], related_phyla[1], candidate), 'root')\n",
    "                phyla_relationship_flag = 1\n",
    "                continue\n",
    "\n",
    "        with cd(f'candidates/iqtree_visualization/%i-%i' % related_phyla):\n",
    "            with open('%s-cluster%s.figTree' % (group_id, cluster_num), 'w') as out:\n",
    "                out.write(visualize_reconstructed_candidate(deepcopy(tmp_tree), related_phyla))\n",
    "\n",
    "        #\n",
    "        # if the flag still is False, we couldn't identify an relationship between phyla\n",
    "        if not phyla_relationship_flag:\n",
    "            descriptions['no clear relationship between phyla'].add(candidate)\n",
    "\n",
    "    return(descriptions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [],
   "source": [
    "fastTree_nested = pkl.load(open('fastTree_nested', 'rb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 256,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1090, 1117)\n",
      "(1090, 200795)\n",
      "(1090, 976)\n",
      "COG0584#0\n",
      "COG3118#0\n",
      "(1090, 1134404)\n",
      "(1090, 1798710)\n",
      "(1117, 200795)\n",
      "(1117, 976)\n",
      "(1117, 1134404)\n",
      "(1117, 1798710)\n",
      "(200795, 976)\n",
      "(200795, 1134404)\n",
      "(200795, 1798710)\n",
      "(976, 1134404)\n",
      "(976, 1798710)\n",
      "(1134404, 1798710)\n"
     ]
    }
   ],
   "source": [
    "descriptions = {}\n",
    "for phyla_pair, relation_description in fastTree_nested.items():\n",
    "#     if 200795 not in phyla_pair or 1117 not in phyla_pair:\n",
    "#         continue\n",
    "    if 1224 in phyla_pair:\n",
    "        continue\n",
    "    \n",
    "    if not os.path.isdir(f'candidates/iqtree_visualization/%i-%i' % phyla_pair):\n",
    "        os.mkdir(f'candidates/iqtree_visualization/%i-%i' % phyla_pair)\n",
    "    \n",
    "    print(phyla_pair)\n",
    "\n",
    "    candidates_to_test = set()\n",
    "    for relation_type, candidates in relation_description.items():\n",
    "        if not re.match('from \\d+ to \\d+', relation_type):\n",
    "            continue\n",
    "            \n",
    "        for candidate in candidates:\n",
    "            candidate_id = candidate.split()[0]\n",
    "            candidates_to_test.add(candidate_id)\n",
    "    \n",
    "    descriptions[phyla_pair] = assess_transfers_between_phyla(phyla_pair, candidates_to_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 257,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "from Chlorobi to Cyanobacteria: 0\n",
      "from Cyanobacteria to Chlorobi: 8\n",
      "    COG0605#1 from Cyanobacteria to Candidatus Thermochlorobacteriaceae\n",
      "    COG0605#1 from Cyanobacteria to Chlorobiaceae\n",
      "    COG5523#0 from Cyanobacteria to Chlorobaculum parvum\n",
      "    COG0003#7 from Cyanobacteria to Chlorobiaceae\n",
      "    COG5523#0 from Cyanobacteria to Chlorobiaceae\n",
      "    COG1518#1 from Cyanobacteria to Chlorobiaceae\n",
      "    COG1351#9 from Cyanobacteria to Chlorobiaceae\n",
      "    COG2867#5 from Cyanobacteria to Chlorobiaceae\n",
      "sisters: 1\n",
      "no clear relationship between phyla: 0\n",
      "\n",
      "from Chlorobi to Chloroflexi: 1\n",
      "    COG0633#1 from Chlorobiaceae to Chloroflexales\n",
      "from Chloroflexi to Chlorobi: 1\n",
      "    COG2738#0 from Chloroflexi to Chlorobiaceae\n",
      "sisters: 0\n",
      "no clear relationship between phyla: 1\n",
      "\n",
      "from Chlorobi to Bacteroidetes: 1\n",
      "    COG1496#1 from Chlorobiales to Bacteroidetes\n",
      "from Bacteroidetes to Chlorobi: 142\n",
      "    COG0498#2 from Bacteroidetes to Chlorobiaceae\n",
      "    COG0142#3 from Bacteroidetes to Chlorobiales\n",
      "    COG1884#2 from Bacteroidetes to Chlorobiaceae\n",
      "    COG0612#0 from Bacteroidetes to Chlorobiaceae\n",
      "    COG0611#1 from Bacteroidetes to Chlorobiaceae\n",
      "    COG0088#3 from Bacteroidetes to Chlorobiales\n",
      "    COG0449#2 from Bacteroidetes to Chlorobiales\n",
      "    COG0489#1 from Bacteroidetes to Chlorobiales\n",
      "    COG0221#2 from Bacteroidetes to Chlorobiaceae\n",
      "    COG4231#0 from Bacteroidia to Chlorobiaceae\n",
      "    COG1884#4 from Bacteroidetes to Chlorobiaceae\n",
      "    COG3808#5 from Bacteroidetes to Chlorobiaceae\n",
      "    COG4288#1 from Bacteroidetes to Chlorobiales\n",
      "    COG1109#0 from Bacteroidetes to Chlorobiales\n",
      "    COG3959#5 from Bacteroidetes to Chlorobiales\n",
      "    COG0621#8 from Bacteroidetes to Chlorobiales\n",
      "    COG0250#4 from Bacteroidetes to Chlorobiaceae\n",
      "    COG1077#3 from Bacteroidetes to Chlorobiales\n",
      "    COG0261#3 from Bacteroidetes to Chlorobiales\n",
      "    COG0249#1 from Bacteroidetes to Chlorobiales\n",
      "    COG4704#1 from Bacteroidetes to Chlorobiaceae\n",
      "    COG2264#0 from Bacteroidetes to Chlorobiales\n",
      "    COG0201#3 from Bacteroidetes to Chlorobiales\n",
      "    COG0475#0 from Bacteroidetes to Chlorobiaceae\n",
      "    COG0130#0 from Bacteroidetes to Chlorobiales\n",
      "    COG0809#3 from Bacteroidetes to Chlorobiales\n",
      "    2ZBRA#1 from Bacteroidetes to Chlorobiaceae\n",
      "    COG1314#2 from Bacteroidetes to Chlorobiaceae\n",
      "    COG1055#4 from Bacteroidetes to Chlorobiaceae\n",
      "    COG3630#0 from Bacteroidetes to Chlorobiaceae\n",
      "    COG0694#0 from Bacteroidetes to Chlorobiales\n",
      "    2Z7RU#0 from Bacteroidetes to Chlorobiaceae\n",
      "    COG0460#4 from Bacteroidetes to Chlorobiaceae\n",
      "    COG0014#0 from Bacteroidetes to Chlorobiaceae\n",
      "    COG1995#0 from Bacteroidetes to Chlorobiales\n",
      "    COG2262#1 from Bacteroidetes to Chlorobiales\n",
      "    COG2825#3 from Bacteroidetes to Chlorobiales\n",
      "    COG0185#1 from Bacteroidetes to Chlorobiales\n",
      "    COG0612#0 from Bacteroidetes to Chloroherpeton thalassium\n",
      "    COG1770#0 from Bacteroidetes to Chlorobiaceae\n",
      "    COG0319#4 from Bacteroidetes to Chlorobiales\n",
      "    COG0098#3 from Bacteroidetes to Chlorobiales\n",
      "    COG0805#2 from Bacteroidetes to Chlorobiales\n",
      "    COG1508#1 from Bacteroidetes to Chlorobiales\n",
      "    COG1088#0 from Bacteroidetes to Chlorobiaceae\n",
      "    COG1053#4 from Bacteroidetes to Chlorobiaceae\n",
      "    COG2771#5 from Bacteroidetes to Chlorobiaceae\n",
      "    COG1702#6 from Bacteroidetes to Chlorobiales\n",
      "    COG0229#0 from Bacteroidetes to Chloroherpeton thalassium\n",
      "    COG2120#2 from Bacteroidetes to Chloroherpeton thalassium\n",
      "    COG2062#0 from Bacteroidetes to Chlorobiaceae\n",
      "    COG0094#3 from Bacteroidetes to Chlorobiales\n",
      "    COG0750#4 from Bacteroidetes to Chlorobiales\n",
      "    COG1209#5 from Bacteroidetes to Chlorobiaceae\n",
      "    COG0221#2 from Bacteroidetes to Chlorobium phaeobacteroides\n",
      "    COG0224#1 from Bacteroidetes to Chlorobiales\n",
      "    COG1010#0 from Bacteroidetes to Chlorobiaceae\n",
      "    COG3118#1 from Bacteroidetes to Chlorobiaceae\n",
      "    COG0083#1 from Bacteroidetes to Chlorobiales\n",
      "    COG0229#2 from Bacteroidetes to Chlorobiaceae\n",
      "    32SXV#0 from Bacteroidetes to Chlorobiaceae\n",
      "    COG5587#1 from Bacteroidetes to Candidatus Thermochlorobacteriaceae\n",
      "    COG0237#6 from Bacteroidetes to Chlorobiales\n",
      "    COG1797#0 from Bacteroidetes to Chlorobiaceae\n",
      "    COG2896#1 from Bacteroidetes to Chlorobiaceae\n",
      "    COG2255#3 from Bacteroidetes to Chlorobiales\n",
      "    COG0742#1 from Bacteroidetes to Chlorobiaceae\n",
      "    COG2120#2 from Bacteroidetes to Chlorobium limicola\n",
      "    COG0072#0 from Bacteroidetes to Chlorobiales\n",
      "    COG0761#0 from Bacteroidetes to Chlorobiales\n",
      "    COG0475#0 from Bacteroidetes to Chlorobaculum\n",
      "    COG0592#3 from Bacteroidetes to Chlorobiales\n",
      "    COG0001#5 from Bacteroidetes to Chlorobiales\n",
      "    COG2815#4 from Bacteroidetes to Chlorobiaceae\n",
      "    COG4192#1 from Bacteroidetes to Chlorobiaceae\n",
      "    COG1022#0 from Bacteroidetes to Chloroherpeton thalassium\n",
      "    COG2875#1 from Bacteroidetes to Chlorobiaceae\n",
      "    COG0479#6 from Bacteroidetes to Chlorobiaceae\n",
      "    COG0092#2 from Bacteroidetes to Chlorobiales\n",
      "    COG1410#1 from Bacteroidetes to Chlorobiaceae\n",
      "    COG0317#13 from Bacteroidetes to Chlorobiales\n",
      "    COG0229#0 from Bacteroidetes to Candidatus Thermochlorobacteriaceae\n",
      "    COG4775#6 from Bacteroidetes to Chlorobiales\n",
      "    COG0594#2 from Bacteroidetes to Chlorobiaceae\n",
      "    COG2242#1 from Bacteroidetes to Chlorobiaceae\n",
      "    COG0593#1 from Bacteroidetes to Chlorobiales\n",
      "    COG0779#4 from Bacteroidetes to Chlorobiales\n",
      "    COG0635#2 from Bacteroidetes to Chlorobiaceae\n",
      "    COG0546#1 from Bacteroidetes to Chlorobiaceae\n",
      "    COG2195#2 from Bacteroidetes to Chlorobiaceae\n",
      "    COG2243#0 from Bacteroidetes to Chlorobiaceae\n",
      "    COG4108#1 from Bacteroidetes to Chlorobiaceae\n",
      "    32RN1#0 from Bacteroidetes to Chlorobiales\n",
      "    COG1884#2 from Bacteroidetes to Chloroherpeton thalassium\n",
      "    COG1883#3 from Bacteroidetes to Chlorobiaceae\n",
      "    32SXV#0 from Bacteroidetes to Prosthecochloris aestuarii\n",
      "    COG0105#1 from Bacteroidetes to Chlorobiales\n",
      "    COG1678#1 from Bacteroidetes to Chlorobiales\n",
      "    COG1808#3 from Bacteroidetes to Chlorobiaceae\n",
      "    COG2827#0 from Bacteroidetes to Chlorobiaceae\n",
      "    COG2896#1 from Bacteroidetes to Candidatus Thermochlorobacteriaceae\n",
      "    COG2073#4 from Bacteroidetes to Chlorobiaceae\n",
      "    COG1209#5 from Bacteroidetes to Chloroherpeton thalassium\n",
      "    COG1259#0 from Bacteroidetes to Chlorobiales\n",
      "    COG0013#5 from Bacteroidetes to Chlorobiales\n",
      "    COG0229#0 from Bacteroidetes to Chlorobiaceae\n",
      "    COG0090#4 from Bacteroidetes to Chlorobiales\n",
      "    COG0342#2 from Bacteroidetes to Chlorobiales\n",
      "    COG0230#1 from Bacteroidetes to Chlorobiaceae\n",
      "    COG0465#0 from Bacteroidetes to Chlorobiales\n",
      "    COG5587#1 from Bacteroidetes to Chlorobiaceae\n",
      "    COG0861#4 from Bacteroidetes to Chlorobiaceae\n",
      "    COG0499#6 from Bacteroidia to Chlorobiaceae\n",
      "    COG0860#4 from Bacteroidetes to Chlorobiaceae\n",
      "    COG2827#0 from Bacteroidetes to Chlorobium phaeobacteroides\n",
      "    COG1541#1 from Bacteroidia to Chlorobiaceae\n",
      "    COG0608#3 from Bacteroidetes to Chlorobiales\n",
      "    COG1014#0 from Bacteroidia to Chlorobiaceae\n",
      "    COG0027#4 from Bacteroidetes to Chlorobiaceae\n",
      "    COG3911#1 from Bacteroidetes to Chlorobiaceae\n",
      "    COG4327#2 from Bacteroidetes to Chlorobiaceae\n",
      "    COG0621#5 from Bacteroidetes to Chlorobiales\n",
      "    COG0127#4 from Bacteroidetes to Chlorobiales\n",
      "    COG0051#1 from Bacteroidetes to Chlorobiales\n",
      "    COG0498#2 from Bacteroidetes to Candidatus Thermochlorobacteriaceae\n",
      "    COG0029#1 from Bacteroidetes to Chlorobiales\n",
      "    COG1399#6 from Bacteroidetes to Chlorobiaceae\n",
      "    COG0250#2 from Bacteroidetes to Chlorobiaceae\n",
      "    COG4147#1 from Bacteroidetes to Chloroherpeton thalassium\n",
      "    COG0378#0 from Bacteroidetes to Chlorobiaceae\n",
      "    COG0509#0 from Bacteroidetes to Candidatus Thermochlorobacteriaceae\n",
      "    COG3264#1 from Bacteroidetes to Chlorobiaceae\n",
      "    COG0258#0 from Bacteroidetes to Chlorobiales\n",
      "    COG0205#4 from Bacteroidia to Chlorobiaceae\n",
      "    COG0263#0 from Bacteroidetes to Chlorobiaceae\n",
      "    COG0646#2 from Bacteroidetes to Chlorobiaceae\n",
      "    COG2062#0 from Bacteroidetes to Chloroherpeton thalassium\n",
      "    COG0460#4 from Bacteroidetes to Candidatus Thermochlorobacteriaceae\n",
      "    COG2082#0 from Bacteroidetes to Chlorobiaceae\n",
      "    COG1137#3 from Bacteroidetes to Chlorobiales\n",
      "    COG0336#3 from Bacteroidetes to Chlorobiales\n",
      "    COG1884#4 from Bacteroidetes to Chloroherpeton thalassium\n",
      "sisters: 13\n",
      "no clear relationship between phyla: 18\n",
      "\n",
      "from Cyanobacteria to Chloroflexi: 21\n",
      "    COG4663#1 from crown Cyanobacteria to stem Chloroflexi\n",
      "    COG4978#0 from Cyanobacteria to Ktedonobacter racemifer\n",
      "    COG2343#0 from Cyanobacteria to Chloroflexus\n",
      "    COG1666#2 from crown Cyanobacteria to stem Chloroflexi\n",
      "    COG5000#0 from Cyanobacteria to Herpetosiphon aurantiacus\n",
      "    COG1075#1 from Cyanobacteria to Chloroflexales\n",
      "    COG1941#3 from Cyanobacteria to Chloroflexales\n",
      "    COG1894#0 from Cyanobacteria to Chloroflexales\n",
      "    COG2343#0 from Cyanobacteria to Oscillochloris trichoides\n",
      "    COG4467#0 from Cyanobacteria to Chloroflexales\n",
      "    COG5000#0 from Cyanobacteria to Chloroflexales\n",
      "    COG1517#2 from Cyanobacteria to Chloroflexia\n",
      "    COG5000#0 from crown Cyanobacteria to stem Chloroflexi\n",
      "    COG1905#1 from Cyanobacteria to Chloroflexales\n",
      "    COG2981#4 from Cyanobacteria to Chloroflexales\n",
      "    COG5635#3 from crown Cyanobacteria to stem Chloroflexi\n",
      "    COG1941#3 from Cyanobacteria to Roseiflexus\n",
      "    COG0115#2 from Cyanobacteria to Thermomicrobia\n",
      "    COG2343#0 from Cyanobacteria to Roseiflexus\n",
      "    COG4118#7 from Cyanobacteria to Chloroflexales\n",
      "    COG5000#0 from Cyanobacteria to Sphaerobacter thermophilus\n",
      "from Chloroflexi to Cyanobacteria: 19\n",
      "    COG2452#1 from crown Chloroflexi to stem Cyanobacteria\n",
      "    COG3379#4 from Chloroflexales to Cyanobacteria\n",
      "    COG0637#4 from crown Chloroflexi to stem Cyanobacteria\n",
      "    COG0345#1 from crown Chloroflexi to stem Cyanobacteria\n",
      "    COG0554#3 from crown Chloroflexi to stem Cyanobacteria\n",
      "    COG0001#4 from crown Chloroflexi to stem Cyanobacteria\n",
      "    COG3259#2 from crown Chloroflexi to stem Cyanobacteria\n",
      "    COG5207#0 from Ktedonobacter racemifer to Cyanobacteria\n",
      "    COG1704#1 from crown Chloroflexi to stem Cyanobacteria\n",
      "    COG1336#1 from Chloroflexales to Cyanobacteria\n",
      "    COG1351#7 from Chloroflexia to Cyanobacteria\n",
      "    COG0499#0 from crown Chloroflexi to stem Cyanobacteria\n",
      "    COG1086#8 from crown Chloroflexi to stem Cyanobacteria\n",
      "    COG0354#0 from crown Chloroflexi to stem Cyanobacteria\n",
      "    COG4978#0 from Herpetosiphon aurantiacus to Cyanobacteria\n",
      "    COG0049#2 from crown Chloroflexi to stem Cyanobacteria\n",
      "    COG1304#5 from crown Chloroflexi to stem Cyanobacteria\n",
      "    COG1554#0 from crown Chloroflexi to stem Cyanobacteria\n",
      "    COG1177#0 from crown Chloroflexi to stem Cyanobacteria\n",
      "sisters: 2\n",
      "no clear relationship between phyla: 2\n",
      "\n",
      "from Cyanobacteria to Bacteroidetes: 14\n",
      "    COG4185#5 from crown Cyanobacteria to stem Bacteroidetes\n",
      "    COG0116#1 from crown Cyanobacteria to stem Bacteroidetes\n",
      "    COG5428#2 from crown Cyanobacteria to stem Bacteroidetes\n",
      "    COG1853#0 from crown Cyanobacteria to stem Bacteroidetes\n",
      "    COG1533#0 from crown Cyanobacteria to stem Bacteroidetes\n",
      "    COG2886#0 from Cyanobacteria to Spirosoma panaciterrae\n",
      "    COG2886#0 from Cyanobacteria to Haliscomenobacter hydrossis\n",
      "    COG5499#2 from Cyanobacteria to Echinicola pacifica\n",
      "    COG5378#0 from crown Cyanobacteria to stem Bacteroidetes\n",
      "    COG2886#0 from crown Cyanobacteria to stem Bacteroidetes\n",
      "    COG0602#4 from crown Cyanobacteria to stem Bacteroidetes\n",
      "    32YEY#3 from Cyanobacteria to Flavobacteriales\n",
      "    COG3563#3 from crown Cyanobacteria to stem Bacteroidetes\n",
      "    COG0701#7 from crown Cyanobacteria to stem Bacteroidetes\n",
      "from Bacteroidetes to Cyanobacteria: 38\n",
      "    COG2515#2 from crown Bacteroidetes to stem Cyanobacteria\n",
      "    COG0590#3 from crown Bacteroidetes to stem Cyanobacteria\n",
      "    COG0415#0 from Bacteroidetes to Gloeobacter\n",
      "    COG4123#4 from crown Bacteroidetes to stem Cyanobacteria\n",
      "    COG3483#1 from crown Bacteroidetes to stem Cyanobacteria\n",
      "    30823#0 from crown Bacteroidetes to stem Cyanobacteria\n",
      "    COG2603#0 from crown Bacteroidetes to stem Cyanobacteria\n",
      "    COG0530#0 from crown Bacteroidetes to stem Cyanobacteria\n",
      "    32T2H#1 from crown Bacteroidetes to stem Cyanobacteria\n",
      "    COG0010#1 from crown Bacteroidetes to stem Cyanobacteria\n",
      "    COG4299#5 from crown Bacteroidetes to stem Cyanobacteria\n",
      "    2ZB0G#0 from crown Bacteroidetes to stem Cyanobacteria\n",
      "    COG2068#1 from crown Bacteroidetes to stem Cyanobacteria\n",
      "    COG3118#6 from crown Bacteroidetes to stem Cyanobacteria\n",
      "    2ZI89#0 from crown Bacteroidetes to stem Cyanobacteria\n",
      "    COG5499#2 from crown Bacteroidetes to stem Cyanobacteria\n",
      "    COG1187#1 from crown Bacteroidetes to stem Cyanobacteria\n",
      "    COG5279#6 from crown Bacteroidetes to stem Cyanobacteria\n",
      "    32RWC#3 from crown Bacteroidetes to stem Cyanobacteria\n",
      "    COG1539#2 from crown Bacteroidetes to stem Cyanobacteria\n",
      "    COG0415#0 from crown Bacteroidetes to stem Cyanobacteria\n",
      "    COG3329#2 from crown Bacteroidetes to stem Cyanobacteria\n",
      "    2Z8XX#0 from crown Bacteroidetes to stem Cyanobacteria\n",
      "    3314C#1 from crown Bacteroidetes to stem Cyanobacteria\n",
      "    331PP#0 from crown Bacteroidetes to stem Cyanobacteria\n",
      "    3314C#1 from Flavobacteriaceae to Cyanobacteria\n",
      "    COG0330#6 from crown Bacteroidetes to stem Cyanobacteria\n",
      "    COG0857#1 from crown Bacteroidetes to stem Cyanobacteria\n",
      "    32ZS2#2 from crown Bacteroidetes to stem Cyanobacteria\n",
      "    COG2942#1 from crown Bacteroidetes to stem Cyanobacteria\n",
      "    COG0182#1 from crown Bacteroidetes to stem Cyanobacteria\n",
      "    COG3034#2 from crown Bacteroidetes to stem Cyanobacteria\n",
      "    2Z951#0 from crown Bacteroidetes to stem Cyanobacteria\n",
      "    2ZTX9#0 from crown Bacteroidetes to stem Cyanobacteria\n",
      "    COG1099#3 from crown Bacteroidetes to stem Cyanobacteria\n",
      "    COG0115#3 from crown Bacteroidetes to stem Cyanobacteria\n",
      "    COG2886#3 from crown Bacteroidetes to stem Cyanobacteria\n",
      "    COG0216#4 from crown Bacteroidetes to stem Cyanobacteria\n",
      "sisters: 8\n",
      "no clear relationship between phyla: 4\n",
      "\n",
      "from Chloroflexi to Bacteroidetes: 0\n",
      "from Bacteroidetes to Chloroflexi: 4\n",
      "    COG0092#2 from crown Bacteroidetes to stem Chloroflexi\n",
      "    COG5563#3 from Bacteroidetes to Herpetosiphon aurantiacus\n",
      "    COG0861#3 from Bacteroidetes to Chloroflexales\n",
      "    COG0861#3 from Bacteroidetes to Herpetosiphon aurantiacus\n",
      "sisters: 1\n",
      "no clear relationship between phyla: 0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "should_ignore  = {1224, 1134404, 1798710}\n",
    "for pair, description in descriptions.items():\n",
    "#     if 200795 not in pair or 1117 not in pair:\n",
    "#         continue\n",
    "    if should_ignore.intersection(pair):\n",
    "        continue\n",
    "\n",
    "    for key, value in description.items():\n",
    "        try:\n",
    "            donor_phyla, recipient_phyla = re.match('from (\\d+) to (\\d+)', key).groups()\n",
    "        except AttributeError:\n",
    "            print(key+':', len(value))\n",
    "            continue\n",
    "        \n",
    "        donor_phyla, recipient_phyla = ncbi.translate_to_names([int(donor_phyla), int(recipient_phyla)])\n",
    "        print(f'from {donor_phyla} to {recipient_phyla}:', len(value))\n",
    "\n",
    "        for detail in value:\n",
    "            candidate, donor, recipient = re.match('(\\S+)(?: from (\\d+) to (\\d+))?', detail).groups()\n",
    "            if donor and recipient:\n",
    "                donor, recipient = ncbi.translate_to_names([int(donor), int(recipient)])\n",
    "                print(f'    {candidate} from {donor} to {recipient}')\n",
    "            else:\n",
    "                print(f'    {candidate} from crown {donor_phyla} to stem {recipient_phyla}')\n",
    "    print()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 240,
   "metadata": {},
   "outputs": [],
   "source": [
    "def visualize_reconstructed_candidate(tree, phyla_pair):\n",
    "\n",
    "    \n",
    "    out = \"#NEXUS\\nbegin taxa;\\n\\tdimensions ntax=%i;\\n\\ttaxlabels\\n\" %len(tree)\n",
    "\n",
    "    count               = 0\n",
    "    internal_node_names = {}\n",
    "    for node in tree.traverse():\n",
    "        if node.is_leaf():\n",
    "            taxid, locus_tag = re.match('(\\d+?)\\.(\\S+)', node.name, re.M).groups()\n",
    "#             taxid, locus_tag = node.name.split('.')\n",
    "            try:\n",
    "                lineage = {j: i for i, j in ncbi.get_rank(ncbi.get_lineage(taxid)).items()}\n",
    "            except ValueError:\n",
    "                out += '\\t%s\\n' %(node.name)\n",
    "                continue\n",
    "            else:\n",
    "                lineage_names = ncbi.get_taxid_translator(lineage.values())\n",
    "\n",
    "#             print(lineage)\n",
    "            out += '\\t%s ' %(node.name)\n",
    "            comment = []\n",
    "            for rank in ['class', 'phylum', 'order', 'family', 'species']:\n",
    "                if rank in lineage:\n",
    "                    comment.append('tax_%s=\"%s\"' %(rank, lineage_names[lineage[rank]]))\n",
    "            if 'phylum' not in lineage:\n",
    "                pass\n",
    "            elif lineage['phylum'] == phyla_pair[0]:\n",
    "                comment.append('!color=#00ff00')\n",
    "            elif lineage['phylum'] == phyla_pair[1]:\n",
    "                comment.append('!color=#ff0000')\n",
    "                            \n",
    "            out += '[&%s]\\n' %' '.join(comment)\n",
    "\n",
    "        else:\n",
    "            internal_node_names['node_%i_' % count] = '[&node_name=%s,ufboot=%.2f, alrt=%.2f]' % (deepcopy(node.name), node.ufboot, node.alrt)\n",
    "            node.name = 'node_%i_' % count\n",
    "            count += 1\n",
    "\n",
    "    newick_text = tree.write(format=1)\n",
    "    for tmp_name, full_name in internal_node_names.items():\n",
    "        newick_text = newick_text.replace(tmp_name, full_name)\n",
    "        \n",
    "    out += ';\\nend;\\n'\n",
    "    out += 'begin trees;\\n\\ttree tree_1 = [&R] %s\\nend;' %newick_text\n",
    "\n",
    "    return(out)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
